<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="CUHK2025">
<title>Unlocking single cell spatial omics analyses with scdney • CUHK2025</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Unlocking single cell spatial omics analyses with scdney">
<meta property="og:description" content="CUHK2025">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">CUHK2025</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/CUHK_workshop_v2.html">Unlocking single cell spatial omics analyses with scdney</a>
    <a class="dropdown-item" href="../articles/CUHK_workshop_v3.html">Unlocking single cell spatial omics analyses with scdney</a>
    <a class="dropdown-item" href="../articles/CUHK_workshop_v4.html">Unlocking single cell spatial omics analyses with scdney</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Unlocking single cell spatial omics analyses with scdney</h1>
            
            <h4 data-toc-skip class="date">06/13/2025</h4>
      
      
      <div class="d-none name"><code>CUHK_workshop_v3.qmd</code></div>
    </div>

    
    
<style>
.question {
  padding: 1em;
  background: lightcyan;
  color: black;
  border-radius: 10px;
}
</style>
<p><strong>Presenting authors</strong></p>
<p>Daniel Kim<span class="math inline">\(^{1,3}\)</span>, Lijia Yu<span class="math inline">\(^{1,2}\)</span>, Daniel Kim<span class="math inline">\(^{1,3}\)</span>, Andrew Zhang<span class="math inline">\(^{1,5}\)</span>, Jean Yang<span class="math inline">\(^{1,2,4}\)</span>.</p>
<p><strong>Contributors</strong></p>
<p>Yue Cao<span class="math inline">\(^{1,2}\)</span>, Lijia Yu<span class="math inline">\(^{1,2}\)</span>, Andy Tran<span class="math inline">\(^{1,2}\)</span>, Daniel Kim<span class="math inline">\(^{1,3}\)</span>, Dario Strbenac<span class="math inline">\(^{1,2}\)</span>, Nicholas Robertson<span class="math inline">\(^{1}\)</span>, Helen Fu<span class="math inline">\(^{3,4}\)</span>, Jean Yang<span class="math inline">\(^{1,2,4}\)</span>.</p>
<p><span class="math inline">\(^1\)</span> Sydney Precision Data Science
Centre, University of Sydney, Australia<br><span class="math inline">\(^2\)</span> School of Mathematics and
Statistics, University of Sydney, Australia<br><span class="math inline">\(^3\)</span> Faculty of Medicine and Health,
University of Sydney, Australia<br><span class="math inline">\(^4\)</span> Charles Perkins Centre,
University of Sydney, Australia<br><span class="math inline">\(^5\)</span> School of Computer Science,
University of Sydney, Australia</p>
<p><br> Contact: jean.yang@sydney.edu.au</p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The emergence of high-resolution spatial omics technologies has
revolutionized our ability to map cellular ecosystems in situ. This
workshop explores approaches in multi-sample spatial data analysis. We
cover end-to-end workflow considerations—from experimental design and QC
to spatial feature interpretation—with case studies in disease
prediction.</p>
<div class="panel-tabset">



<ul class="nav nav-tabs" id="NA" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#our-dataset" id="our-dataset-tab" type="button" role="tab" aria-controls="our-dataset" aria-selected="true" class="active nav-link">Our dataset</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#assumed-knowledge" id="assumed-knowledge-tab" type="button" role="tab" aria-controls="assumed-knowledge" aria-selected="false" class="nav-link">Assumed knowledge</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#learning-objectives" id="learning-objectives-tab" type="button" role="tab" aria-controls="learning-objectives" aria-selected="false" class="nav-link">Learning objectives</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="our-dataset" role="tabpanel" aria-labelledby="our-dataset-tab">

<p>A subset of the widely-known METABRIC breast cancer cohort has
recently been analyzed using imaging mass cytometry: <a href="https://www.nature.com/articles/s43018-020-0026-6" class="external-link">Imaging Mass
Cytometry and Multiplatform Genomics Define the Phenogenomic Landscape
of Breast Cancer</a>. Patient clinical data was sourced from the
Supplementary Table 5 of <a href="https://www.nature.com/articles/s41586-019-1007-8" class="external-link">Dynamics of
Breast-cancer Relapse Reveal Late-recurring ER-positive Genomic
Subgroups</a>.</p>
</div>
<div class="tab-pane" id="assumed-knowledge" role="tabpanel" aria-labelledby="assumed-knowledge-tab">

<ul>
<li>Experience with R.</li>
<li>Familiarity with the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment
class</a>.</li>
<li>Basic knowledge in <a href="https://bioconductor.org/books/release/OSCA/index.html" class="external-link">single
cell data analysis</a>. You can access our <a href="https://github.com/SydneyBioX/scdney#scdney-workshops-series" class="external-link">previous
workshops</a> for a quick tutorial.</li>
<li>Ability to install all required R packages, please check
<code>sessionInfo</code> at the end of this document to ensure you are
using the correct versions.</li>
<li>Familiarity with our previous workshop vignette on <a href="https://sydneybiox.github.io/BIS2019_SC/" class="external-link">Introduction to Single
Cell RNA-seq Analysis</a>.</li>
</ul>
</div>
<div class="tab-pane" id="learning-objectives" role="tabpanel" aria-labelledby="learning-objectives-tab">

<ul>
<li>Describe and visualise spatial omics datasets.</li>
<li>Understand approaches for quality control of spatial proteomics
data.</li>
<li>Calculate spatial statistics at the cell-type level using
<code>scFeatures</code>.</li>
<li>Perform multi-view disease outcome prediction with the package
<code>ClassifyR</code>.</li>
<li>Develop an understanding of:
<ul>
<li>evaluation of classification and survival models .</li>
<li>evaluate cohort heterogeneity given a survival model.</li>
</ul>
</li>
<li>Explore various strategies for disease outcome prognosis using
spatial omics data.</li>
</ul>
</div>
</div>
</div>
<div class="question">
<p><strong>Our question of interest</strong></p>
<p>We want to know if the risk of recurrence in the METABRIC breast
cancer cohort can be accurately estimated to inform how aggressively
they need to be treated.</p>
</div>
</div>
<div class="section level2">
<h2 id="part-1-exploring-the-data">Part 1: Exploring the data<a class="anchor" aria-label="anchor" href="#part-1-exploring-the-data"></a>
</h2>
<div class="section level3">
<h3 id="initial-data-exploration">1.1: Initial data exploration<a class="anchor" aria-label="anchor" href="#initial-data-exploration"></a>
</h3>
<p>At the start of any analysis pipeline, it is often good to explore
the data to get a sense of the structure and its complexity. Let’s
explore the data to answer the questions below:</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>How many features and observations are there in the data and what do
the features represent?</li>
<li>What covariates are in our data?</li>
<li>Given our question of interest, what variable would be our outcome
variable?</li>
</ol>
</div>
<p><br></p>
<div class="panel-tabset">



<ul class="nav nav-tabs" id="NA" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#sce-object" id="sce-object-tab" type="button" role="tab" aria-controls="sce-object" aria-selected="true" class="active nav-link">SCE object</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#data-overview" id="data-overview-tab" type="button" role="tab" aria-controls="data-overview" aria-selected="false" class="nav-link">Data overview</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#covariates" id="covariates-tab" type="button" role="tab" aria-controls="covariates" aria-selected="false" class="nav-link">Covariates</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="sce-object" role="tabpanel" aria-labelledby="sce-object-tab">

<p>First, we take a quick look at the structure of the SingleCell
Experiment object.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"../../data/breastCancer.RData"</span><span class="op">)</span></span>
<span><span class="va">data_sce</span> <span class="op">=</span> <span class="va">IMC</span></span>
<span></span>
<span><span class="co"># Structure of our data</span></span>
<span><span class="va">data_sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 38 76307 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(38): HH3_total CK19 ... H3K27me3 CK5</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(76307): MB-0002:345:93 MB-0002:345:107 ... MB-0663:394:577</span></span>
<span><span class="co">##   MB-0663:394:578</span></span>
<span><span class="co">## colData names(17): file_id metabricId ... x_cord y_cord</span></span>
<span><span class="co">## reducedDimNames(1): UMAP</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Metadata</span></span>
<span><span class="co">#DT::datatable(data.frame(colData(data_sce)))</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                      file_id metabricId core_id ImageNumber ObjectNumber</span></span>
<span><span class="co">## MB-0002:345:93  MB0002_1_345    MB-0002       1         345           93</span></span>
<span><span class="co">## MB-0002:345:107 MB0002_1_345    MB-0002       1         345          107</span></span>
<span><span class="co">## MB-0002:345:113 MB0002_1_345    MB-0002       1         345          113</span></span>
<span><span class="co">## MB-0002:345:114 MB0002_1_345    MB-0002       1         345          114</span></span>
<span><span class="co">## MB-0002:345:125 MB0002_1_345    MB-0002       1         345          125</span></span>
<span><span class="co">## MB-0002:345:127 MB0002_1_345    MB-0002       1         345          127</span></span>
<span><span class="co">##                 Fibronectin Location_Center_X Location_Center_Y SOM_nodes</span></span>
<span><span class="co">## MB-0002:345:93    0.4590055          179.7088          27.90110       130</span></span>
<span><span class="co">## MB-0002:345:107   0.1176471          193.0588          29.29412       131</span></span>
<span><span class="co">## MB-0002:345:113   0.2512903          162.1935          28.83871        83</span></span>
<span><span class="co">## MB-0002:345:114   0.4788444          198.3111          29.15556       128</span></span>
<span><span class="co">## MB-0002:345:125   6.1901112          262.8889          31.26667         7</span></span>
<span><span class="co">## MB-0002:345:127   0.6897432          136.0946          32.98649       130</span></span>
<span><span class="co">##                 pg_cluster    description   region      ki67</span></span>
<span><span class="co">## MB-0002:345:93          54       HR- CK7- region_1  low_ki67</span></span>
<span><span class="co">## MB-0002:345:107         54       HR- CK7- region_4  low_ki67</span></span>
<span><span class="co">## MB-0002:345:113         28    HRlow CKlow region_1 high_ki67</span></span>
<span><span class="co">## MB-0002:345:114         54       HR- CK7- region_4 high_ki67</span></span>
<span><span class="co">## MB-0002:345:125         11 Myofibroblasts region_2  low_ki67</span></span>
<span><span class="co">## MB-0002:345:127         54       HR- CK7- region_1  low_ki67</span></span>
<span><span class="co">##                               celltype  sample   x_cord   y_cord</span></span>
<span><span class="co">## MB-0002:345:93        HR- CK7--region1 MB-0002 179.7088 27.90110</span></span>
<span><span class="co">## MB-0002:345:107       HR- CK7--region4 MB-0002 193.0588 29.29412</span></span>
<span><span class="co">## MB-0002:345:113    HRlow CKlow-region1 MB-0002 162.1935 28.83871</span></span>
<span><span class="co">## MB-0002:345:114       HR- CK7--region4 MB-0002 198.3111 29.15556</span></span>
<span><span class="co">## MB-0002:345:125 Myofibroblasts-region2 MB-0002 262.8889 31.26667</span></span>
<span><span class="co">## MB-0002:345:127       HR- CK7--region1 MB-0002 136.0946 32.98649</span></span></code></pre>
</div>
<div class="tab-pane" id="data-overview" role="tabpanel" aria-labelledby="data-overview-tab">

<p>Here, we take a quick look at what our rows and columns represent and
the dimensions of our data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Glimpse the first few observations </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "MB-0002:345:93"  "MB-0002:345:107" "MB-0002:345:113" "MB-0002:345:114"</span></span>
<span><span class="co">## [5] "MB-0002:345:125" "MB-0002:345:127"</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Glimpse the first few features</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "HH3_total" "CK19"      "CK8_18"    "Twist"     "CD68"      "CK14"</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># How many features and observations are in our dataset?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]    38 76307</span></span></code></pre>
</div>
<div class="tab-pane" id="covariates" role="tabpanel" aria-labelledby="covariates-tab">

<p>In addition to the proteomics data, it’s important to understand what
other covariates, such as clinical variables, are in our dataset. These
can help us in answering our question(s) of interest or formulate new
questions. For example, we can’t explore the association between smoking
status and breast cancer severity if the variable for smoking status
doesn’t exist in our data.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Explore covariates</span></span>
<span></span>
<span><span class="co"># DT::datatable(clinical)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "file_id"           "metabricId"        "core_id"          </span></span>
<span><span class="co">##  [4] "ImageNumber"       "ObjectNumber"      "Fibronectin"      </span></span>
<span><span class="co">##  [7] "Location_Center_X" "Location_Center_Y" "SOM_nodes"        </span></span>
<span><span class="co">## [10] "pg_cluster"        "description"       "region"           </span></span>
<span><span class="co">## [13] "ki67"              "celltype"          "sample"           </span></span>
<span><span class="co">## [16] "x_cord"            "y_cord"</span></span></code></pre>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="is-this-a-complex-dataset">1.2: Is this a complex dataset?<a class="anchor" aria-label="anchor" href="#is-this-a-complex-dataset"></a>
</h3>
<p>Now that we have a basic idea of what our data looks like, we can
look at it in more detail. While initial data exploration reveals
fundamental patterns, deeper examination is very helpful. As it serves
two critical purposes: first, to detect anomalies or biases requiring
remediation, and second, to inform our choice of analytical methods
tailored to the biological questions at hand.</p>
<div class="panel-tabset">




<ul class="nav nav-tabs" id="NA" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#missing-values" id="missing-values-tab" type="button" role="tab" aria-controls="missing-values" aria-selected="true" class="active nav-link">Missing values</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#imbalance" id="imbalance-tab" type="button" role="tab" aria-controls="imbalance" aria-selected="false" class="nav-link">Imbalance</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#relationships" id="relationships-tab" type="button" role="tab" aria-controls="relationships" aria-selected="false" class="nav-link">Relationships</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#visualise-cells" id="visualise-cells-tab" type="button" role="tab" aria-controls="visualise-cells" aria-selected="false" class="nav-link">Visualise cells</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="missing-values" role="tabpanel" aria-labelledby="missing-values-tab">

<p>Examining the proportion of missing values in a dataset is crucial
for ensuring the accuracy and validity of data analysis. Missing values
can significantly impact the reliability of statistical results,
potentially leading to biased conclusions and reduced statistical power.
Here, we plot the proportion of missing values for each of our clinical
variables.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># What is the proportion of missing values in our clinical data?</span></span>
<span><span class="fu">gg_miss_var</span><span class="op">(</span><span class="va">clinical</span>, show_pct <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="tab-pane" id="imbalance" role="tabpanel" aria-labelledby="imbalance-tab">

<p>Looking at imbalance in data is crucial because it can lead to biased
models and inaccurate predictions, especially in classification tasks.
Imbalance occurs when one class is significantly underrepresented
compared to others. This can cause models to be overly influenced by the
majority class, leading to poor performance on the minority class. Here,
we tabularise some of the relevant clinical variables and plot the
distribution of the <code>Time to Recurrence-Free Survival</code>
variable.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estrogen receptor status</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">clinical</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">ER.Status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_bar</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#0099B4"</span>, color <span class="op">=</span> <span class="st">"white"</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Count\n"</span>, </span>
<span>       x <span class="op">=</span> <span class="st">"\nER status"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Histological type</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">clinical</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Histological.Type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_bar</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#fc6203"</span>, color <span class="op">=</span> <span class="st">"white"</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Count\n"</span>, </span>
<span>       x <span class="op">=</span> <span class="st">"\nHistological Type"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># eventRFS: "Event in Recurrence-Free Survival."It indicates whether the event has occurred.#</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">clinical</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">eventRFS</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_bar</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#40dbb2"</span>, color <span class="op">=</span> <span class="st">"white"</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Count\n"</span>, </span>
<span>       x <span class="op">=</span> <span class="st">"\nEvent in Recurrence-Free Survival"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-5-3.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># timeRFS: "Time to Recurrence-Free Survival." It is the time period until recurrence occurs. </span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">clinical</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">timeRFS</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#de9921"</span>, color <span class="op">=</span> <span class="st">"white"</span>, alpha <span class="op">=</span> <span class="fl">0.8</span>, bins<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Frequency\n"</span>, </span>
<span>       x <span class="op">=</span> <span class="st">"\ntimeRFS"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-5-4.png" width="700"></p>
</div>
<div class="tab-pane" id="relationships" role="tabpanel" aria-labelledby="relationships-tab">

<p>Here, we explore the distribution of the outcomes and variables in
the meta-data. We use cross-tabulation to examine the following
variables: Surgery vs death, ER status, and Grade.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Stage and death</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">$</span><span class="va">Breast.Surgery</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">Death</span>, useNA <span class="op">=</span> <span class="st">"ifany"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">##                    </span></span>
<span><span class="co">##                      0  1</span></span>
<span><span class="co">##   BREAST CONSERVING 38 14</span></span>
<span><span class="co">##   MASTECTOMY        16  6</span></span>
<span><span class="co">##   &lt;NA&gt;               2  1</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ER status and grade</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">$</span><span class="va">ER.Status</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">Grade</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      </span></span>
<span><span class="co">##        1  2  3</span></span>
<span><span class="co">##   neg  0  0 11</span></span>
<span><span class="co">##   pos 13 30 17</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># "Number of individuals based on Grade</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">$</span><span class="va">Grade</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">Death</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    </span></span>
<span><span class="co">##      0  1</span></span>
<span><span class="co">##   1 12  2</span></span>
<span><span class="co">##   2 25  5</span></span>
<span><span class="co">##   3 17 11</span></span></code></pre>
</div>
<div class="tab-pane" id="visualise-cells" role="tabpanel" aria-labelledby="visualise-cells-tab">

<p>To assess potential batch effects and sample-specific clustering
patterns, we visualize the cells in our data colored by the sample of
origin. This qualitative inspection provides an intuitive first
assessment of any batch effects in our data integration. If there is
batch effects then we’d need to apply batch correction to resolve this
issue.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#data_sce &lt;- runUMAP(data_sce, scale=TRUE)</span></span>
<span></span>
<span><span class="co"># With the UMAP function we can highlight by meta data of interest</span></span>
<span><span class="co"># Here we highlight the UMAP by sample ID</span></span>
<span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">data_sce</span>, colour_by <span class="op">=</span> <span class="st">"metabricId"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-7-1.png" width="1152"></p>
</div>
</div>
</div>
<div class="question">
<p><em>Discussion</em></p>
<ul>
<li>Are there any anomalies or biases that might need to be corrected to
ensure our analysis is robust?</li>
<li>Given our question of interest and the characteristics of our data,
are there any particular analytic techniques that would be appropriate?
Are there any that would not be appropriate?</li>
</ul>
</div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="part-2-quality-control">Part 2: Quality control<a class="anchor" aria-label="anchor" href="#part-2-quality-control"></a>
</h2>
<p>Here, we explore key approaches for evaluating the quality of IMC
data. A robust assessment should consider multiple factors, including
the density of marker expression, marker correlations, and their
co-expression. Something we might want to think about is what
characeristics might indicate whether a sample(s) is low/high
quality?</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_counts</span> <span class="op">&lt;-</span> <span class="va">IMC</span><span class="op">@</span><span class="va">colData</span> <span class="op">|&gt;</span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">metabricId</span>, name <span class="op">=</span> <span class="st">"cell_count"</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="co"># Count rows per metabricId</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">cell_count</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">IMC</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">IMC</span><span class="op">@</span><span class="va">colData</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Location_Center_X"</span>,<span class="st">"Location_Center_Y"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="va">cell_type_mapping</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"B cells"</span> <span class="op">=</span> <span class="st">"B cell"</span>,</span>
<span>  <span class="st">"T cells"</span> <span class="op">=</span> <span class="st">"T cell"</span>,</span>
<span>  <span class="st">"Macrophages Vim+ CD45low"</span> <span class="op">=</span> <span class="st">"Macrophage"</span>,</span>
<span>  <span class="st">"Macrophages Vim+ Slug-"</span> <span class="op">=</span> <span class="st">"Macrophage"</span>,</span>
<span>  <span class="st">"Macrophages Vim+ Slug+"</span> <span class="op">=</span> <span class="st">"Macrophage"</span>,</span>
<span></span>
<span>  <span class="st">"Endothelial"</span> <span class="op">=</span> <span class="st">"Endothelial"</span>,</span>
<span></span>
<span>  <span class="st">"Fibroblasts"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span>  <span class="st">"Fibroblasts CD68+"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span>  <span class="st">"Myofibroblasts"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span>  <span class="st">"Vascular SMA+"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span></span>
<span>  <span class="st">"Myoepithelial"</span> <span class="op">=</span> <span class="st">"Myoepithelial"</span>,</span>
<span></span>
<span>  <span class="st">"HR+ CK7-"</span> <span class="op">=</span> <span class="st">"Tumor HR+"</span>,</span>
<span>  <span class="st">"HR+ CK7- Ki67+"</span> <span class="op">=</span> <span class="st">"Tumor HR+"</span>,</span>
<span>  <span class="st">"HR+ CK7- Slug+"</span> <span class="op">=</span> <span class="st">"Tumor HR+"</span>,</span>
<span></span>
<span>  <span class="st">"HR- CK7+"</span> <span class="op">=</span> <span class="st">"Tumor HR-"</span>,</span>
<span>  <span class="st">"HR- CK7-"</span> <span class="op">=</span> <span class="st">"Tumor HR-"</span>,</span>
<span>  <span class="st">"HR- Ki67+"</span> <span class="op">=</span> <span class="st">"Tumor HR-"</span>,</span>
<span>  <span class="st">"HR- CKlow CK5+"</span> <span class="op">=</span> <span class="st">"Tumor HR-"</span>,</span>
<span></span>
<span>  <span class="st">"HRlow CKlow"</span> <span class="op">=</span> <span class="st">"Tumor HR-low/mixed"</span>,</span>
<span>  <span class="st">"HER2+"</span> <span class="op">=</span> <span class="st">"Tumor HER2+"</span>,</span>
<span>  <span class="st">"Basal CKlow"</span> <span class="op">=</span> <span class="st">"Tumor Basal-like"</span>,</span>
<span>  <span class="st">"Hypoxia"</span> <span class="op">=</span> <span class="st">"Tumor Hypoxic"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert IMC description to higher-level categories</span></span>
<span><span class="va">IMC</span><span class="op">$</span><span class="va">high_level_category</span> <span class="op">&lt;-</span> <span class="fu">recode</span><span class="op">(</span><span class="va">IMC</span><span class="op">$</span><span class="va">description</span>, <span class="op">!</span><span class="op">!</span><span class="op">!</span><span class="va">cell_type_mapping</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">imc.spe</span><span class="op">=</span><span class="va">IMC</span><span class="op">[</span>,<span class="va">IMC</span><span class="op">$</span><span class="va">metabricId</span><span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span><span class="va">cell_counts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,<span class="op">]</span><span class="op">$</span><span class="va">metabricId</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Convert IMC description to higher-level categories</span></span>
<span><span class="va">imc.spe</span><span class="op">$</span><span class="va">high_level_category</span> <span class="op">&lt;-</span> <span class="fu">recode</span><span class="op">(</span><span class="va">imc.spe</span><span class="op">$</span><span class="va">description</span>, <span class="op">!</span><span class="op">!</span><span class="op">!</span><span class="va">cell_type_mapping</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu">plot_marker_densities</span><span class="op">(</span><span class="va">imc.spe</span>, sample_id_col <span class="op">=</span> <span class="st">"metabricId"</span>,</span>
<span>                                markers_to_plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD20"</span>,<span class="st">"CD3"</span>,<span class="st">"CD68"</span>,<span class="st">"vWF_CD31"</span>,<span class="st">"Vimentin"</span>,<span class="st">"HER2"</span><span class="op">)</span>,</span>
<span>                               assay_name <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="qc-for-the-panel-markers">2.1: QC for the panel / markers<a class="anchor" aria-label="anchor" href="#qc-for-the-panel-markers"></a>
</h3>
<p>Here, we want to assess the quality of the markers: it is desirable
to see at least two peaks, which would indicate the existence of cell
type specific markers in our data.</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Can we see any cell type specific markers, which are they?</li>
<li>Are there any non-cell type specific markers?</li>
</ol>
</div>
<p><br></p>
<div class="panel-tabset">







<ul class="nav nav-tabs" id="NA" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">plots</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"#### "</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">plots</span><span class="op">)</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="st">"\n"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">plots</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\n\n"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#cd20" id="cd20-tab" type="button" role="tab" aria-controls="cd20" aria-selected="false" class="nav-link">CD20</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#cd3" id="cd3-tab" type="button" role="tab" aria-controls="cd3" aria-selected="false" class="nav-link">CD3</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#cd68" id="cd68-tab" type="button" role="tab" aria-controls="cd68" aria-selected="false" class="nav-link">CD68</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#vwf_cd31" id="vwf_cd31-tab" type="button" role="tab" aria-controls="vwf_cd31" aria-selected="false" class="nav-link">vWF_CD31</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#vimentin" id="vimentin-tab" type="button" role="tab" aria-controls="vimentin" aria-selected="false" class="nav-link">Vimentin</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#her2" id="her2-tab" type="button" role="tab" aria-controls="her2" aria-selected="false" class="nav-link">HER2</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab"></div>
<div class="tab-pane" id="cd20" role="tabpanel" aria-labelledby="cd20-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="tab-pane" id="cd3" role="tabpanel" aria-labelledby="cd3-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-9-2.png" width="700"></p>
</div>
<div class="tab-pane" id="cd68" role="tabpanel" aria-labelledby="cd68-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-9-3.png" width="700"></p>
</div>
<div class="tab-pane" id="vwf_cd31" role="tabpanel" aria-labelledby="vwf_cd31-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-9-4.png" width="700"></p>
</div>
<div class="tab-pane" id="vimentin" role="tabpanel" aria-labelledby="vimentin-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-9-5.png" width="700"></p>
</div>
<div class="tab-pane" id="her2" role="tabpanel" aria-labelledby="her2-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-9-6.png" width="700"></p>
</div>
</div>
</div>
<div class="question">
<p><em>Discussion</em></p>
<ul>
<li>Do we have a sufficient amount of cell type specific markers?</li>
<li>Is the panel of genes sufficient for our study?</li>
</ul>
</div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="qc-of-individual-samples">2.2: QC of individual samples<a class="anchor" aria-label="anchor" href="#qc-of-individual-samples"></a>
</h3>
<p>Here we plot the correlation of different markers for a subset of
samples. This helps us determine whether a sample has low/high
contamination of markers.</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Which pairs of markers do we expect to be correlated/not
correlated.</li>
<li>Is there any evidence of marker contamination for a given
sample?</li>
</ol>
</div>
<p><br></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coexp_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../../data/coexp_df.rds"</span><span class="op">)</span></span>
<span><span class="va">marker_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD20"</span>, <span class="st">"CD3"</span>, <span class="st">"CD68"</span>, <span class="st">"vWF_CD31"</span>, <span class="st">"Vimentin"</span>, <span class="st">"HER2"</span><span class="op">)</span></span>
<span> </span>
<span><span class="co"># Proportion heatmaps</span></span>
<span><span class="va">heatmaps_prop</span> <span class="op">&lt;-</span> <span class="fu">plot_pairwise_heatmaps_per_sample</span><span class="op">(</span><span class="va">coexp_df</span>, <span class="va">marker_list</span>, stat <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span></span></code></pre></div>
<div class="panel-tabset">














































































<ul class="nav nav-tabs" id="NA" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">heatmaps_prop</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"#### "</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">heatmaps_prop</span><span class="op">)</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="st">"\n"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">heatmaps_prop</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\n\n"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0002" id="mb-0002-tab" type="button" role="tab" aria-controls="mb-0002" aria-selected="false" class="nav-link">MB-0002</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0064" id="mb-0064-tab" type="button" role="tab" aria-controls="mb-0064" aria-selected="false" class="nav-link">MB-0064</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0128" id="mb-0128-tab" type="button" role="tab" aria-controls="mb-0128" aria-selected="false" class="nav-link">MB-0128</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0130" id="mb-0130-tab" type="button" role="tab" aria-controls="mb-0130" aria-selected="false" class="nav-link">MB-0130</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0132" id="mb-0132-tab" type="button" role="tab" aria-controls="mb-0132" aria-selected="false" class="nav-link">MB-0132</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0136" id="mb-0136-tab" type="button" role="tab" aria-controls="mb-0136" aria-selected="false" class="nav-link">MB-0136</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0138" id="mb-0138-tab" type="button" role="tab" aria-controls="mb-0138" aria-selected="false" class="nav-link">MB-0138</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0142" id="mb-0142-tab" type="button" role="tab" aria-controls="mb-0142" aria-selected="false" class="nav-link">MB-0142</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0145" id="mb-0145-tab" type="button" role="tab" aria-controls="mb-0145" aria-selected="false" class="nav-link">MB-0145</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0154" id="mb-0154-tab" type="button" role="tab" aria-controls="mb-0154" aria-selected="false" class="nav-link">MB-0154</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0168" id="mb-0168-tab" type="button" role="tab" aria-controls="mb-0168" aria-selected="false" class="nav-link">MB-0168</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0175" id="mb-0175-tab" type="button" role="tab" aria-controls="mb-0175" aria-selected="false" class="nav-link">MB-0175</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0180" id="mb-0180-tab" type="button" role="tab" aria-controls="mb-0180" aria-selected="false" class="nav-link">MB-0180</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0192" id="mb-0192-tab" type="button" role="tab" aria-controls="mb-0192" aria-selected="false" class="nav-link">MB-0192</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0201" id="mb-0201-tab" type="button" role="tab" aria-controls="mb-0201" aria-selected="false" class="nav-link">MB-0201</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0208" id="mb-0208-tab" type="button" role="tab" aria-controls="mb-0208" aria-selected="false" class="nav-link">MB-0208</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0225" id="mb-0225-tab" type="button" role="tab" aria-controls="mb-0225" aria-selected="false" class="nav-link">MB-0225</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0227" id="mb-0227-tab" type="button" role="tab" aria-controls="mb-0227" aria-selected="false" class="nav-link">MB-0227</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0232" id="mb-0232-tab" type="button" role="tab" aria-controls="mb-0232" aria-selected="false" class="nav-link">MB-0232</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0240" id="mb-0240-tab" type="button" role="tab" aria-controls="mb-0240" aria-selected="false" class="nav-link">MB-0240</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0245" id="mb-0245-tab" type="button" role="tab" aria-controls="mb-0245" aria-selected="false" class="nav-link">MB-0245</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0246" id="mb-0246-tab" type="button" role="tab" aria-controls="mb-0246" aria-selected="false" class="nav-link">MB-0246</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0249" id="mb-0249-tab" type="button" role="tab" aria-controls="mb-0249" aria-selected="false" class="nav-link">MB-0249</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0252" id="mb-0252-tab" type="button" role="tab" aria-controls="mb-0252" aria-selected="false" class="nav-link">MB-0252</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0255" id="mb-0255-tab" type="button" role="tab" aria-controls="mb-0255" aria-selected="false" class="nav-link">MB-0255</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0256" id="mb-0256-tab" type="button" role="tab" aria-controls="mb-0256" aria-selected="false" class="nav-link">MB-0256</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0258" id="mb-0258-tab" type="button" role="tab" aria-controls="mb-0258" aria-selected="false" class="nav-link">MB-0258</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0260" id="mb-0260-tab" type="button" role="tab" aria-controls="mb-0260" aria-selected="false" class="nav-link">MB-0260</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0263" id="mb-0263-tab" type="button" role="tab" aria-controls="mb-0263" aria-selected="false" class="nav-link">MB-0263</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0275" id="mb-0275-tab" type="button" role="tab" aria-controls="mb-0275" aria-selected="false" class="nav-link">MB-0275</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0308" id="mb-0308-tab" type="button" role="tab" aria-controls="mb-0308" aria-selected="false" class="nav-link">MB-0308</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0318" id="mb-0318-tab" type="button" role="tab" aria-controls="mb-0318" aria-selected="false" class="nav-link">MB-0318</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0320" id="mb-0320-tab" type="button" role="tab" aria-controls="mb-0320" aria-selected="false" class="nav-link">MB-0320</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0347" id="mb-0347-tab" type="button" role="tab" aria-controls="mb-0347" aria-selected="false" class="nav-link">MB-0347</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0351" id="mb-0351-tab" type="button" role="tab" aria-controls="mb-0351" aria-selected="false" class="nav-link">MB-0351</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0367" id="mb-0367-tab" type="button" role="tab" aria-controls="mb-0367" aria-selected="false" class="nav-link">MB-0367</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0377" id="mb-0377-tab" type="button" role="tab" aria-controls="mb-0377" aria-selected="false" class="nav-link">MB-0377</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0383" id="mb-0383-tab" type="button" role="tab" aria-controls="mb-0383" aria-selected="false" class="nav-link">MB-0383</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0394" id="mb-0394-tab" type="button" role="tab" aria-controls="mb-0394" aria-selected="false" class="nav-link">MB-0394</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0397" id="mb-0397-tab" type="button" role="tab" aria-controls="mb-0397" aria-selected="false" class="nav-link">MB-0397</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0400" id="mb-0400-tab" type="button" role="tab" aria-controls="mb-0400" aria-selected="false" class="nav-link">MB-0400</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0401" id="mb-0401-tab" type="button" role="tab" aria-controls="mb-0401" aria-selected="false" class="nav-link">MB-0401</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0405" id="mb-0405-tab" type="button" role="tab" aria-controls="mb-0405" aria-selected="false" class="nav-link">MB-0405</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0410" id="mb-0410-tab" type="button" role="tab" aria-controls="mb-0410" aria-selected="false" class="nav-link">MB-0410</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0420" id="mb-0420-tab" type="button" role="tab" aria-controls="mb-0420" aria-selected="false" class="nav-link">MB-0420</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0439" id="mb-0439-tab" type="button" role="tab" aria-controls="mb-0439" aria-selected="false" class="nav-link">MB-0439</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0451" id="mb-0451-tab" type="button" role="tab" aria-controls="mb-0451" aria-selected="false" class="nav-link">MB-0451</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0454" id="mb-0454-tab" type="button" role="tab" aria-controls="mb-0454" aria-selected="false" class="nav-link">MB-0454</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0455" id="mb-0455-tab" type="button" role="tab" aria-controls="mb-0455" aria-selected="false" class="nav-link">MB-0455</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0461" id="mb-0461-tab" type="button" role="tab" aria-controls="mb-0461" aria-selected="false" class="nav-link">MB-0461</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0463" id="mb-0463-tab" type="button" role="tab" aria-controls="mb-0463" aria-selected="false" class="nav-link">MB-0463</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0467" id="mb-0467-tab" type="button" role="tab" aria-controls="mb-0467" aria-selected="false" class="nav-link">MB-0467</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0470" id="mb-0470-tab" type="button" role="tab" aria-controls="mb-0470" aria-selected="false" class="nav-link">MB-0470</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0475" id="mb-0475-tab" type="button" role="tab" aria-controls="mb-0475" aria-selected="false" class="nav-link">MB-0475</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0480" id="mb-0480-tab" type="button" role="tab" aria-controls="mb-0480" aria-selected="false" class="nav-link">MB-0480</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0481" id="mb-0481-tab" type="button" role="tab" aria-controls="mb-0481" aria-selected="false" class="nav-link">MB-0481</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0487" id="mb-0487-tab" type="button" role="tab" aria-controls="mb-0487" aria-selected="false" class="nav-link">MB-0487</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0498" id="mb-0498-tab" type="button" role="tab" aria-controls="mb-0498" aria-selected="false" class="nav-link">MB-0498</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0508" id="mb-0508-tab" type="button" role="tab" aria-controls="mb-0508" aria-selected="false" class="nav-link">MB-0508</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0516" id="mb-0516-tab" type="button" role="tab" aria-controls="mb-0516" aria-selected="false" class="nav-link">MB-0516</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0518" id="mb-0518-tab" type="button" role="tab" aria-controls="mb-0518" aria-selected="false" class="nav-link">MB-0518</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0520" id="mb-0520-tab" type="button" role="tab" aria-controls="mb-0520" aria-selected="false" class="nav-link">MB-0520</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0531" id="mb-0531-tab" type="button" role="tab" aria-controls="mb-0531" aria-selected="false" class="nav-link">MB-0531</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0536" id="mb-0536-tab" type="button" role="tab" aria-controls="mb-0536" aria-selected="false" class="nav-link">MB-0536</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0571" id="mb-0571-tab" type="button" role="tab" aria-controls="mb-0571" aria-selected="false" class="nav-link">MB-0571</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0584" id="mb-0584-tab" type="button" role="tab" aria-controls="mb-0584" aria-selected="false" class="nav-link">MB-0584</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0591" id="mb-0591-tab" type="button" role="tab" aria-controls="mb-0591" aria-selected="false" class="nav-link">MB-0591</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0596" id="mb-0596-tab" type="button" role="tab" aria-controls="mb-0596" aria-selected="false" class="nav-link">MB-0596</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0604" id="mb-0604-tab" type="button" role="tab" aria-controls="mb-0604" aria-selected="false" class="nav-link">MB-0604</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0605" id="mb-0605-tab" type="button" role="tab" aria-controls="mb-0605" aria-selected="false" class="nav-link">MB-0605</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0628" id="mb-0628-tab" type="button" role="tab" aria-controls="mb-0628" aria-selected="false" class="nav-link">MB-0628</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0635" id="mb-0635-tab" type="button" role="tab" aria-controls="mb-0635" aria-selected="false" class="nav-link">MB-0635</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0636" id="mb-0636-tab" type="button" role="tab" aria-controls="mb-0636" aria-selected="false" class="nav-link">MB-0636</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0642" id="mb-0642-tab" type="button" role="tab" aria-controls="mb-0642" aria-selected="false" class="nav-link">MB-0642</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0661" id="mb-0661-tab" type="button" role="tab" aria-controls="mb-0661" aria-selected="false" class="nav-link">MB-0661</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0662" id="mb-0662-tab" type="button" role="tab" aria-controls="mb-0662" aria-selected="false" class="nav-link">MB-0662</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#mb-0663" id="mb-0663-tab" type="button" role="tab" aria-controls="mb-0663" aria-selected="false" class="nav-link">MB-0663</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab"></div>
<div class="tab-pane" id="mb-0002" role="tabpanel" aria-labelledby="mb-0002-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0064" role="tabpanel" aria-labelledby="mb-0064-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-2.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0128" role="tabpanel" aria-labelledby="mb-0128-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-3.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0130" role="tabpanel" aria-labelledby="mb-0130-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-4.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0132" role="tabpanel" aria-labelledby="mb-0132-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-5.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0136" role="tabpanel" aria-labelledby="mb-0136-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-6.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0138" role="tabpanel" aria-labelledby="mb-0138-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-7.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0142" role="tabpanel" aria-labelledby="mb-0142-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-8.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0145" role="tabpanel" aria-labelledby="mb-0145-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-9.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0154" role="tabpanel" aria-labelledby="mb-0154-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-10.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0168" role="tabpanel" aria-labelledby="mb-0168-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-11.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0175" role="tabpanel" aria-labelledby="mb-0175-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-12.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0180" role="tabpanel" aria-labelledby="mb-0180-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-13.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0192" role="tabpanel" aria-labelledby="mb-0192-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-14.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0201" role="tabpanel" aria-labelledby="mb-0201-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-15.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0208" role="tabpanel" aria-labelledby="mb-0208-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-16.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0225" role="tabpanel" aria-labelledby="mb-0225-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-17.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0227" role="tabpanel" aria-labelledby="mb-0227-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-18.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0232" role="tabpanel" aria-labelledby="mb-0232-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-19.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0240" role="tabpanel" aria-labelledby="mb-0240-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-20.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0245" role="tabpanel" aria-labelledby="mb-0245-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-21.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0246" role="tabpanel" aria-labelledby="mb-0246-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-22.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0249" role="tabpanel" aria-labelledby="mb-0249-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-23.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0252" role="tabpanel" aria-labelledby="mb-0252-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-24.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0255" role="tabpanel" aria-labelledby="mb-0255-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-25.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0256" role="tabpanel" aria-labelledby="mb-0256-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-26.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0258" role="tabpanel" aria-labelledby="mb-0258-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-27.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0260" role="tabpanel" aria-labelledby="mb-0260-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-28.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0263" role="tabpanel" aria-labelledby="mb-0263-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-29.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0275" role="tabpanel" aria-labelledby="mb-0275-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-30.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0308" role="tabpanel" aria-labelledby="mb-0308-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-31.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0318" role="tabpanel" aria-labelledby="mb-0318-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-32.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0320" role="tabpanel" aria-labelledby="mb-0320-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-33.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0347" role="tabpanel" aria-labelledby="mb-0347-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-34.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0351" role="tabpanel" aria-labelledby="mb-0351-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-35.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0367" role="tabpanel" aria-labelledby="mb-0367-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-36.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0377" role="tabpanel" aria-labelledby="mb-0377-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-37.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0383" role="tabpanel" aria-labelledby="mb-0383-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-38.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0394" role="tabpanel" aria-labelledby="mb-0394-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-39.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0397" role="tabpanel" aria-labelledby="mb-0397-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-40.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0400" role="tabpanel" aria-labelledby="mb-0400-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-41.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0401" role="tabpanel" aria-labelledby="mb-0401-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-42.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0405" role="tabpanel" aria-labelledby="mb-0405-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-43.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0410" role="tabpanel" aria-labelledby="mb-0410-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-44.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0420" role="tabpanel" aria-labelledby="mb-0420-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-45.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0439" role="tabpanel" aria-labelledby="mb-0439-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-46.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0451" role="tabpanel" aria-labelledby="mb-0451-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-47.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0454" role="tabpanel" aria-labelledby="mb-0454-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-48.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0455" role="tabpanel" aria-labelledby="mb-0455-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-49.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0461" role="tabpanel" aria-labelledby="mb-0461-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-50.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0463" role="tabpanel" aria-labelledby="mb-0463-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-51.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0467" role="tabpanel" aria-labelledby="mb-0467-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-52.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0470" role="tabpanel" aria-labelledby="mb-0470-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-53.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0475" role="tabpanel" aria-labelledby="mb-0475-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-54.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0480" role="tabpanel" aria-labelledby="mb-0480-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-55.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0481" role="tabpanel" aria-labelledby="mb-0481-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-56.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0487" role="tabpanel" aria-labelledby="mb-0487-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-57.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0498" role="tabpanel" aria-labelledby="mb-0498-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-58.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0508" role="tabpanel" aria-labelledby="mb-0508-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-59.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0516" role="tabpanel" aria-labelledby="mb-0516-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-60.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0518" role="tabpanel" aria-labelledby="mb-0518-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-61.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0520" role="tabpanel" aria-labelledby="mb-0520-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-62.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0531" role="tabpanel" aria-labelledby="mb-0531-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-63.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0536" role="tabpanel" aria-labelledby="mb-0536-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-64.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0571" role="tabpanel" aria-labelledby="mb-0571-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-65.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0584" role="tabpanel" aria-labelledby="mb-0584-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-66.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0591" role="tabpanel" aria-labelledby="mb-0591-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-67.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0596" role="tabpanel" aria-labelledby="mb-0596-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-68.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0604" role="tabpanel" aria-labelledby="mb-0604-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-69.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0605" role="tabpanel" aria-labelledby="mb-0605-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-70.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0628" role="tabpanel" aria-labelledby="mb-0628-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-71.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0635" role="tabpanel" aria-labelledby="mb-0635-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-72.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0636" role="tabpanel" aria-labelledby="mb-0636-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-73.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0642" role="tabpanel" aria-labelledby="mb-0642-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-74.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0661" role="tabpanel" aria-labelledby="mb-0661-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-75.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0662" role="tabpanel" aria-labelledby="mb-0662-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-76.png" width="700"></p>
</div>
<div class="tab-pane" id="mb-0663" role="tabpanel" aria-labelledby="mb-0663-tab">

<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-11-77.png" width="700"></p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="qc-of-study-as-a-whole">2.3: QC of study as a whole<a class="anchor" aria-label="anchor" href="#qc-of-study-as-a-whole"></a>
</h3>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Are there any samples you would like to remove from the data?</li>
<li>Which samples would you like to keep?</li>
</ol>
</div>
<p><br></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_counts</span> <span class="op">&lt;-</span> <span class="va">IMC</span><span class="op">@</span><span class="va">colData</span> <span class="op">|&gt;</span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">metabricId</span>, name <span class="op">=</span> <span class="st">"cell_count"</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="co"># Count rows per metabricId</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">cell_count</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span></span>
<span><span class="va">mean_coexp_per_sample</span> <span class="op">&lt;-</span> <span class="va">coexp_df</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">cell_id</span>, <span class="op">-</span><span class="va">cell_type</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="co"># remove columns not related to prob values</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">metabricId</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_coexp_prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu">c_across</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">merged_df</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">mean_coexp_per_sample</span>, <span class="va">cell_counts</span>, by <span class="op">=</span> <span class="st">"metabricId"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Scatter plot</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">merged_df</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">cell_count</span>, y <span class="op">=</span> <span class="va">mean_coexp_prob</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span>, color <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">labs</span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Cell Count per Sample"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Mean Co-expression Probability"</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Mean Co-expression vs. Cell Count"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="does-my-data-agree-with-literature">2.4: Does my data agree with literature?<a class="anchor" aria-label="anchor" href="#does-my-data-agree-with-literature"></a>
</h3>
<p>Another way to QC our data is to explore whether the DE genes or
associations we find matches current literature. For example, we have a
breast cancer cohort. Breast cancer is very well studied and so it
should be very easy to perform a DE analysis after pseudobulking our
data to see if it makes findings in previous bulk RNA-seq studies of
breast cancer cohorts. We can also do this at the single-cell level and
compare with previous single-cell proteomic/RNA-seq studies of breast
cancer cohorts. We won’t be running the code below. But here is an
example of how one might perform a basic DE analysis using package <a href="https://academic.oup.com/nar/article/43/7/e47/2414268" class="external-link">limma</a>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note: The same approach can be used for pseudobulk or single-cell level data. </span></span>
<span><span class="co"># Just change `logcounts(data_sce)`</span></span>
<span></span>
<span><span class="co"># We want low ki67 to be the reference level (baseline level) </span></span>
<span><span class="va">data_sce</span><span class="op">$</span><span class="va">ki67</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">$</span><span class="va">ki67</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"low_ki67"</span>, <span class="st">"high_ki67"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Here we specify the design matrix. We are specifying that Y are the expression values and X is ki67 status (but it could be any other variable - such as "good" or "bad" prognosis)</span></span>
<span><span class="co"># Y~X: Expression as a function of ki67 status. </span></span>
<span><span class="va">design_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">data_sce</span><span class="op">$</span><span class="va">ki67</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit model</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">lmFit</span><span class="op">(</span><span class="fu">logcounts</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span>, design <span class="op">=</span> <span class="va">design_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Estimate variance and SE of coefficients</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">eBayes</span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tt</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="fl">2</span>, n <span class="op">=</span> <span class="fl">5</span>, adjust.method <span class="op">=</span> <span class="st">"BH"</span>, sort.by <span class="op">=</span> <span class="st">"p"</span><span class="op">)</span></span>
<span><span class="va">tt</span></span></code></pre></div>
<pre><code><span><span class="co">##               logFC   AveExpr         t P.Value adj.P.Val        B</span></span>
<span><span class="co">## HH3_total 0.7141833 2.4738506 146.94273       0         0 9496.599</span></span>
<span><span class="co">## CK19      0.4401674 0.8206783  81.20218       0         0 3151.934</span></span>
<span><span class="co">## CK8_18    1.0394535 1.4776415 107.90431       0         0 5407.860</span></span>
<span><span class="co">## Twist     0.2506436 0.4034097 125.83265       0         0 7183.755</span></span>
<span><span class="co">## CK14      0.1730002 0.2040003  56.21965       0         0 1538.247</span></span></code></pre>
<div class="Discussion">
<p><em>Discussion</em></p>
<ul>
<li>How might be check whether these results make sense? What potential
issues might we find?</li>
</ul>
</div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="part-3-exploring-spatial-data">Part 3: Exploring spatial data<a class="anchor" aria-label="anchor" href="#part-3-exploring-spatial-data"></a>
</h2>
<p>This section assumes pre-existing cell type annotations. For
unannotated data, two primary annotation strategies are available:</p>
<ul>
<li>Unsupervised approach: cluster cells and identify marker genes for
manual annotation.</li>
<li>Supervised approach: transfer labels from reference datasets using
supervised learning/classification tools.</li>
</ul>
<p>For supervised annotation, we recommend <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC7306901/" class="external-link">scClassify</a>,
a robust framework for cell-type classification. Below we visualise the
data with and without spatial information before exploring the data
further.</p>
<div class="section level4">
<h4 id="cell-types">Cell types<a class="anchor" aria-label="anchor" href="#cell-types"></a>
</h4>
<p>Following initial assessment of potential batch effects through
sample-origin visualization, we now examine cell-type-specific
clustering patterns. Distinct, biologically meaningful clusters should
emerge for each annotated cell type. The presence of heterogeneous
clusters containing unrelated cell types may indicate incomplete or
inaccurate cell-type annotations.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">data_sce</span>, colour_by <span class="op">=</span> <span class="st">"description"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="spatial-structure">Spatial structure<a class="anchor" aria-label="anchor" href="#spatial-structure"></a>
</h4>
<p>The advantage with spatial omics is that we can examine the
organisation of the cell-types as it occurs on the tissue slide. Here,
we visualise one of the slides from a patient. We select a particular
patient “MB-0002” and visualise the tissue sample from this patient
using ggplot. Do we see any spatial patterning or does it look randomly
distributed?</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># obtaining the meta data for this patient </span></span>
<span><span class="va">one_sample</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span>, <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span>  <span class="op">==</span> <span class="st">"MB-0002"</span><span class="op">]</span></span>
<span><span class="va">one_sample</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">one_sample</span><span class="op">)</span><span class="op">)</span></span>
<span> </span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">one_sample</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span>, y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">description</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_colour_tableau</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line<span class="op">=</span><span class="fu">element_line</span><span class="op">(</span>color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Original slide"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-15-1.png" width="576"></p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>What kinds of spatial information might be of interest given the
question we’d like to answer?</li>
<li>How might we try to capture these spatial relationships?</li>
</ol>
</div>
<p><br></p>
</div>
<div class="panel-tabset">



<ul class="nav nav-tabs" id="NA" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#spatial-regions" id="spatial-regions-tab" type="button" role="tab" aria-controls="spatial-regions" aria-selected="true" class="active nav-link">Spatial regions</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#across-individuals-optional" id="across-individuals-optional-tab" type="button" role="tab" aria-controls="across-individuals-optional" aria-selected="false" class="nav-link">Across individuals (Optional)</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#specific-regions-optional" id="specific-regions-optional-tab" type="button" role="tab" aria-controls="specific-regions-optional" aria-selected="false" class="nav-link">Specific regions (Optional)</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="spatial-regions" role="tabpanel" aria-labelledby="spatial-regions-tab">

<p>Background: One of the most common questions or analyses in spatial
data is spatial domain detection. “Spatial domains” are regions within a
tissue where cells share similar gene expression profiles and are
physically clustered together. Here, we will use the terminology
“spatial domain” and “regions” interchangeably. Most common analytical
strategies involve spatial clustering, where different methods use
different levels of information, such as gene expression data only, cell
type information, and spatial coordinates.</p>
<p>The strategy we demonstrate here uses the <code>lisaClust</code>
function, which uses cell type information to cluster cells into five
distinct spatial regions. As a case study, we will compare individuals
with good or poor prognosis and examine them graphically to see if any
regions appear to be different between good or poor prognosis. We
define: - Good prognosis as individuals with &gt; 10 years
recurrence-free survival and - Poor prognosis as individuals with &lt; 5
years recurrence-free survival.</p>
<p>Below we visualise the spatial domain (regions) detection result
based on one individual. Here we will use the terminology “spatial
domain” and “regions” interchangeably.</p>
<p>Depending on the number of regions, it may be more useful to
visualise the spatial regions either collectively in a single graph or
separately in multiple graphs. To visualise it in a single graph, the
<code>hatchingPlot()</code> function is used to create hatching patterns
for representating spatial regions and cell-types.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract time to recurrence-free survival</span></span>
<span><span class="va">clinical</span><span class="op">$</span><span class="va">survivalgroup</span> <span class="op">&lt;-</span> <span class="st">"neither"</span></span>
<span></span>
<span><span class="co">## Define poor and good prognosis</span></span>
<span><span class="va">clinical</span><span class="op">$</span><span class="va">survivalgroup</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="va">clinical</span><span class="op">$</span><span class="va">timeRFS</span>  <span class="op">&lt;</span> <span class="fl">5</span><span class="op">*</span> <span class="fl">365</span><span class="op">)</span> <span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"poor"</span> </span>
<span><span class="va">clinical</span><span class="op">$</span><span class="va">survivalgroup</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="va">clinical</span><span class="op">$</span><span class="va">timeRFS</span>  <span class="op">&gt;</span> <span class="fl">10</span><span class="op">*</span> <span class="fl">365</span><span class="op">)</span> <span class="op">]</span> <span class="op">&lt;-</span>  <span class="st">"good"</span></span>
<span></span>
<span><span class="co">## To visualise it in a single graph</span></span>
<span><span class="fu">hatchingPlot</span><span class="op">(</span></span>
<span>  <span class="va">data_sce</span>,</span>
<span>  region <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>  imageID <span class="op">=</span> <span class="st">"metabricId"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"description"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Location_Center_X"</span>, <span class="st">"Location_Center_Y"</span><span class="op">)</span> <span class="op">)</span> </span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-16-1.png" width="576"></p>
<p><br></p>
<p>We have written a small function
<code>draw_region_clustering_result</code> to visualise the data
separately in multiple graphs for the individual
<code>MB-0002</code>.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_region_clustering_result</span><span class="op">(</span><span class="va">data_sce</span> , </span>
<span>                              sample <span class="op">=</span> <span class="st">"metabricId"</span> , </span>
<span>                              selected_sample <span class="op">=</span>  <span class="st">"MB-0002"</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span></code></pre>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-17-1.png" width="960"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span></code></pre>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-17-2.png" width="960"></p>
<div class="question">
<p><em>Discussion</em></p>
<ul>
<li>If we’d like to examine multiple individuals, is there a better
visualise this information?</li>
</ul>
</div>
<p><br></p>
</div>
<div class="tab-pane" id="across-individuals-optional" role="tabpanel" aria-labelledby="across-individuals-optional-tab">

<p>We can better interpret the region output by summarising the
proportion of each cell-type in a region across the individuals. We look
at the composition of cell-types in each region and compare between
prognostic outcomes.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_dotplot</span><span class="op">(</span><span class="va">data_sce</span>,  </span>
<span>             sample <span class="op">=</span> <span class="st">"metabricId"</span> , </span>
<span>             celltype <span class="op">=</span> <span class="st">"description"</span> ,  </span>
<span>             group<span class="op">=</span> <span class="st">"survivalgroup"</span> , </span>
<span>             group_of_interest <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"poor"</span> , <span class="st">"good"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `summarise()` has grouped output by 'Var1', 'Var2'. You can override using the</span></span>
<span><span class="co">## `.groups` argument.</span></span></code></pre>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-18-1.png" width="960"></p>
</div>
<div class="tab-pane" id="specific-regions-optional" role="tabpanel" aria-labelledby="specific-regions-optional-tab">

<p>The number of sub-cell types increase considerably when we want to
add spatial domain (region) information. To enhance clarity and
facilitate understanding, it may be helpful to choose a predetermined
region. The code generates a set of boxplots that enable the comparison
of cell-type proportions between individuals with good and
poor prognosis in <code>region_5</code>.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_selected_region_boxplot</span><span class="op">(</span><span class="va">data_sce</span>, </span>
<span>                             sample <span class="op">=</span> <span class="st">"metabricId"</span> , </span>
<span>                             celltype <span class="op">=</span><span class="st">"description"</span> , </span>
<span>                             group  <span class="op">=</span> <span class="st">"survivalgroup"</span>, </span>
<span>                             group_of_interest <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"poor"</span> , <span class="st">"good"</span><span class="op">)</span>, </span>
<span>                             select_region <span class="op">=</span> <span class="st">"region_5"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-19-1.png" width="1152"></p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="part-4-feature-engineering-with-scfeatures">Part 4: Feature engineering with scFeatures<a class="anchor" aria-label="anchor" href="#part-4-feature-engineering-with-scfeatures"></a>
</h2>
<p>Here, we use scFeatures to generate molecular features for each
individual using the <code>features x cells</code> matrices. These
features are interpretable and can be used for downstream analyses. In
general, <code>scFeatures</code> generates features across six
categories:</p>
<ol style="list-style-type: lower-roman">
<li>Cell-type proportions.</li>
<li>Cell-type specific gene expressions.</li>
<li>Cell-type specific pathway expressions.</li>
<li>Cell-type interaction scores.</li>
<li>Aggregated gene expressions.</li>
<li>Spatial metrics: Nearest neighbour’s correlation, L statistics, and
Moran’s I.</li>
</ol>
<p>The different types of features constructed enable a more
comprehensive multi-view understanding of each individual. By default,
the function will generate a total of 13 different types of features and
are stored as 13 <code>samples x features</code> matrices, one for each
feature type.</p>
<p>In this section, we will examine spatial information from two
perspectives. Utilising spatial domain detection described in part 3, we
select a specific region of interest and create molecular
representations of that region for each individual (Section 4.1).
Second, we will utilise spatial statistics to capture spatial
relationships within the region of interest, such as Moran’s I (Section
4.2)</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">region</span></span>
<span></span>
<span><span class="co"># Define a series of sub-cell-types that is regional specific</span></span>
<span><span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">description</span> , <span class="st">"-"</span> , <span class="va">region</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="how-to-create-molecular-representations-of-individuals-for-an-roi">4.1 How to create molecular representations of individuals for an
ROI?<a class="anchor" aria-label="anchor" href="#how-to-create-molecular-representations-of-individuals-for-an-roi"></a>
</h3>
<p>Here, we can consider regional information and the following code
allows us to create cell-type specific features for each region. We use
the function “paste0” to construct <strong>region-specific sub
cell-types</strong> and name it as <code>celltype</code> in the R object
<code>data_sce</code>. For simplicity, in this workshop, the variable
<code>celltype</code> in the R object <code>data_sce</code> refers to
region-specific sub-cell-types.</p>
<p>There are a total of 13 different types of features (feature_types)
that you can choose to generate. The argument type refers the type of
input data we have. This is either <code>scrna</code> (single-cell
RNA-sequencing data), <code>spatial_p</code> (spatial proteomics data),
or <code>spatial_t</code> (single-cell spatial data). In this section,
we will ignore spatial information and generate non-spatial features,
such as pseudobulking at the sample /cell type levels, overall
expression, cell type proportions etc…</p>
<div class="panel-tabset">




<ul class="nav nav-tabs" id="NA" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#generate-features" id="generate-features-tab" type="button" role="tab" aria-controls="generate-features" aria-selected="true" class="active nav-link">Generate features</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#roi" id="roi-tab" type="button" role="tab" aria-controls="roi" aria-selected="false" class="nav-link">ROI</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#scfeatures-code-optional" id="scfeatures-code-optional-tab" type="button" role="tab" aria-controls="scfeatures-code-optional" aria-selected="false" class="nav-link">scFeatures code (Optional)</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#scfeatures-output-optional" id="scfeatures-output-optional-tab" type="button" role="tab" aria-controls="scfeatures-output-optional" aria-selected="false" class="nav-link">scFeatures output (Optional)</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="generate-features" role="tabpanel" aria-labelledby="generate-features-tab">

<p>Suppose that we are interested in determining the proportion of
cell-types within each region for each individual. It is necessary to
specify <code>type = spatial_p</code> to reflect that we have spatial
proteomics data and <code>feature_types = proportion_raw</code> to
indicate we intend to calculate cell-type proportions for each of the
region-specific cell-types. Suppose we are only interested in the
molecular representation of <code>HR- CK7+</code> within individuals
within region 5.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [A] The next few lines extract specific information from data_sce as input to scFeatures. </span></span>
<span><span class="co">## Extract the expression matrix from data_sce</span></span>
<span><span class="va">IMCmatrix</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract the sample information </span></span>
<span><span class="co">## append the condition to the individuals so we can easily retrieve the individuals condition </span></span>
<span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> </span>
<span><span class="va">cond</span>  <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">metabricId</span><span class="op">)</span>, <span class="op">]</span><span class="op">$</span><span class="va">survivalgroup</span></span>
<span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"_cond_"</span>, <span class="va">cond</span> <span class="op">)</span> </span>
<span></span>
<span><span class="co">## Extract the region-specific sub-cell-types</span></span>
<span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span></span>
<span></span>
<span><span class="co">## Extract the spatial coordinates</span></span>
<span><span class="va">spatialCoords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_X</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### [B] Running scFeatures</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu">scFeatures</span><span class="op">(</span><span class="va">IMCmatrix</span>, </span>
<span>                                sample <span class="op">=</span> <span class="va">sample</span>, </span>
<span>                                celltype <span class="op">=</span> <span class="va">celltype</span>, </span>
<span>                                spatialCoords <span class="op">=</span> <span class="va">spatialCoords</span>,</span>
<span>                                feature_types <span class="op">=</span> <span class="st">"proportion_raw"</span>, </span>
<span>                                type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span></code></pre></div>
</div>
<div class="tab-pane" id="roi" role="tabpanel" aria-labelledby="roi-tab">

<div class="question">
<p><strong>Question</strong></p>
<ol style="list-style-type: decimal">
<li>Are there any regions that are associated with “good” and “poor”
prognosis?</li>
<li>Is this the right way to visualise the results?</li>
</ol>
</div>
<p><br></p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [A] The next few lines extract specific information from data_sce as input to scFeatures. </span></span>
<span><span class="co">## Select the HR- CK7+-region sub-cell-type </span></span>
<span></span>
<span><span class="co"># There are different ways you can use `scFeatures` to generate molecular representations for individuals and it requires the following information for spatial data.</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># data,\</span></span>
<span><span class="co"># sample,</span></span>
<span><span class="co"># X coordinates,</span></span>
<span><span class="co"># Y coordinates,</span></span>
<span><span class="co"># feature_types, and</span></span>
<span><span class="co"># type</span></span>
<span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span>   <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"HR- CK7+-region"</span> , <span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span>, fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">selected_data</span>  <span class="op">&lt;-</span>  <span class="va">IMCmatrix</span><span class="op">[</span>, <span class="va">index</span><span class="op">]</span></span>
<span><span class="va">selected_sample</span> <span class="op">&lt;-</span>  <span class="va">sample</span><span class="op">[</span><span class="va">index</span><span class="op">]</span></span>
<span><span class="va">selected_celltype</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span> <span class="va">index</span><span class="op">]</span> </span>
<span><span class="va">selected_spatialCoords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_X</span><span class="op">[</span><span class="va">index</span><span class="op">]</span>, </span>
<span>                               <span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">[</span><span class="va">index</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### [B] Running scFeatures</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu">scFeatures</span><span class="op">(</span> <span class="va">selected_data</span>, </span>
<span>                                 sample <span class="op">=</span> <span class="va">selected_sample</span>, </span>
<span>                                 celltype <span class="op">=</span> <span class="va">selected_celltype</span>,</span>
<span>                                 spatialCoords <span class="op">=</span> <span class="va">selected_spatialCoords</span>,</span>
<span>                                 feature_types <span class="op">=</span> <span class="st">"proportion_raw"</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span>
<span></span>
<span><span class="co">### [C] Visualize the regional composition makeup for each individual for HR- CK7+ and HR- CK7-</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">proportion_raw</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span></span>
<span><span class="fu">plot_barplot</span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Proportion raw feature"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"proportion\n"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-22-1.png" width="1920"></p>
</div>
<div class="tab-pane" id="scfeatures-code-optional" role="tabpanel" aria-labelledby="scfeatures-code-optional-tab">

<p>The code below enable us to generate all feature types for all
cell-types in a line. Due to limitations with today’s computational
capacity, <strong>Please DO NOT run it in today’s workshop, it will
crash your system</strong>.</p>
<pre><code><span><span class="co"># here, we specify that this is a spatial proteomics data</span></span>
<span><span class="co"># scFeatures support parallel computation to speed up the process </span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu">scFeatures</span><span class="op">(</span><span class="va">IMCmatrix</span>, </span>
<span>                                type <span class="op">=</span> <span class="st">"spatial_p"</span>,</span>
<span>                                sample <span class="op">=</span> <span class="va">sample</span>, </span>
<span>                                celltype <span class="op">=</span> <span class="va">celltype</span>, </span>
<span>                                spatialCoords <span class="op">=</span> <span class="va">spatialCoords</span>,</span>
<span>                                ncores <span class="op">=</span> <span class="fl">32</span><span class="op">)</span></span></code></pre>
</div>
<div class="tab-pane" id="scfeatures-output-optional" role="tabpanel" aria-labelledby="scfeatures-output-optional-tab">

<p>Assuming you have already generated a collection of molecular
representation for individuals, please load the prepared RDS file
<code>scfeatures_result.RDS</code>. Again, you can remind yourself that
all generated feature types are stored in a matrix of
<code>samples x features</code>.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Upload pre-generated RDS file</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../../data/scfeatures_result.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># What are the features and the dimensions of features matrices that we have generated?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">scfeatures_result</span>, <span class="va">dim</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $proportion_raw</span></span>
<span><span class="co">## [1]  77 110</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $proportion_logit</span></span>
<span><span class="co">## [1]  77 110</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $proportion_ratio</span></span>
<span><span class="co">## [1]   77 5995</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_mean_celltype</span></span>
<span><span class="co">## [1]   77 4180</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_prop_celltype</span></span>
<span><span class="co">## [1]   77 4180</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_cor_celltype</span></span>
<span><span class="co">## [1]    77 27958</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_mean_bulk</span></span>
<span><span class="co">## [1] 77 38</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_prop_bulk</span></span>
<span><span class="co">## [1] 77 38</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_cor_bulk</span></span>
<span><span class="co">## [1]  77 703</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $L_stats</span></span>
<span><span class="co">## [1]   77 3842</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $celltype_interaction</span></span>
<span><span class="co">## [1]   77 4393</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $morans_I</span></span>
<span><span class="co">## [1] 77 38</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $nn_correlation</span></span>
<span><span class="co">## [1] 77 38</span></span></code></pre>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="how-can-we-represent-spatial-information-and-relationships-for-a-given-individual">4.2 How can we represent spatial information and relationships for a
given individual?<a class="anchor" aria-label="anchor" href="#how-can-we-represent-spatial-information-and-relationships-for-a-given-individual"></a>
</h3>
<p>We will now look at the spatial statistic output by
<code>scFeatures</code> and qualitively assess whether there is any
association between between these statistics and the “good” and “bad”
prognosis groups.</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>What kind of information would spatial statistics provide?</li>
<li>Are there any differences in the distribution of spatial statistics
between the “good” and “bad” prognosis groups?</li>
</ol>
</div>
<p><br></p>
<div class="panel-tabset">



<ul class="nav nav-tabs" id="NA" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#nearest-neighbour-correlation" id="nearest-neighbour-correlation-tab" type="button" role="tab" aria-controls="nearest-neighbour-correlation" aria-selected="true" class="active nav-link">Nearest neighbour correlation</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l-statistics" id="l-statistics-tab" type="button" role="tab" aria-controls="l-statistics" aria-selected="false" class="nav-link">L statistics</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#morans-i" id="morans-i-tab" type="button" role="tab" aria-controls="morans-i" aria-selected="false" class="nav-link">Moran’s I</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="nearest-neighbour-correlation" role="tabpanel" aria-labelledby="nearest-neighbour-correlation-tab">

<p>This metric measures the correlation of proteins/genes between a cell
and its nearest neighbour cell.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">nn_correlation</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line<span class="op">=</span><span class="fu">element_line</span><span class="op">(</span>color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Nearest neighbour correlation feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
</div>
<div class="tab-pane" id="l-statistics" role="tabpanel" aria-labelledby="l-statistics-tab">

<p>The L function is a spatial statistic used to assess the spatial
distribution of cell-types. It assesses the significance of cell-cell
interactions, by calculating the density of a cell-type with other
cell-types within a certain radius. High values indicate spatial
association (co-localisation), low values indicate spatial avoidance.For
clarity, we plot the L statistics for a subset of regions.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">L_stats</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^T cells-region4_HR"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span><span class="op">==</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span>,<span class="va">idx</span><span class="op">]</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line<span class="op">=</span><span class="fu">element_line</span><span class="op">(</span>color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"L statistics"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
</div>
<div class="tab-pane" id="morans-i" role="tabpanel" aria-labelledby="morans-i-tab">

<p>Moran’s I is a spatial autocorrelation statistic based on both
location and values. It quantifies whether similar values tend to occur
near each other or are dispersed.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">morans_I</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line<span class="op">=</span><span class="fu">element_line</span><span class="op">(</span>color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Moran's I"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="part-5-disease-classification-with-classifyr">Part 5: Disease classification with ClassifyR<a class="anchor" aria-label="anchor" href="#part-5-disease-classification-with-classifyr"></a>
</h2>
<p>Recurrence risk estimation is a fundamental concern in medical
research, particularly in the context of patient survival analysis. In
this section, we will estimate recurrence risk using the molecular
representation of individuals generated from <code>scFeatures</code> to
build a survival model. We will use classifyR to build the survival
model. The patient outcome is time-to-event, so, by default, ClassifyR
will use Cox proportional hazards ranking to choose a set of features
and also Cox proportional hazards to predict risk scores. We will also
demonstrate other available models in <code>ClassifyR</code>.</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol style="list-style-type: decimal">
<li>Are spatial features globally informative in predicting
survival?</li>
<li>If not, for which individuals is it important in predicting
survival?</li>
</ol>
</div>
<p><br></p>
<div class="panel-tabset">



<ul class="nav nav-tabs" id="NA" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#building-a-survival-model-optional" id="building-a-survival-model-optional-tab" type="button" role="tab" aria-controls="building-a-survival-model-optional" aria-selected="true" class="active nav-link">Building a survival model (Optional)</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#model-performance-optional" id="model-performance-optional-tab" type="button" role="tab" aria-controls="model-performance-optional" aria-selected="false" class="nav-link">Model performance (Optional)</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#does-spatial-information-matter" id="does-spatial-information-matter-tab" type="button" role="tab" aria-controls="does-spatial-information-matter" aria-selected="false" class="nav-link">Does spatial information matter?</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="building-a-survival-model-optional" role="tabpanel" aria-labelledby="building-a-survival-model-optional-tab">

<p>Recall in the previous section that we have stored the 13 matrices of
different feature types in a list. Instead of manually retrieving each
matrix from the list to build separate models, classifyR can directly
take a list of matrices as an input and run a repeated cross-validation
model on each matrix individually. Below, we run 5 repeats of 5-fold
cross-validation. A high score indicates prognosis of a worse outcome
than a lower risk score. Although we have provided the code below, to
save time, <strong>just load the prepared RDS file
<code>classifyr_result_IMC.rds</code></strong> and we will focus on the
interpretation in this workshop.</p>
<pre><code><span><span class="co"># We use the following variables:     </span></span>
<span><span class="co"># timeRFS: "Time to Recurrence-Free Survival." It is the time period until recurrence occurs.    </span></span>
<span><span class="co"># eventRFS: "Event in Recurrence-Free Survival."It indicates whether the event has occurred.    </span></span>
<span><span class="co"># Breast.Tumour.Laterality: Laterality of tumors, eg, whether the tumor is located in left or right.     </span></span>
<span><span class="co"># ER.Status: Whether the tumor is ER positive or ER negative.    </span></span>
<span><span class="co"># Inferred.Menopausal.State: of the patient.    </span></span>
<span><span class="co"># Grade: of the tumor.   </span></span>
<span><span class="co"># Size: of the tumor.   </span></span>
<span></span>
<span><span class="va">usefulFeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Breast.Tumour.Laterality"</span>, <span class="st">"ER.Status"</span>, <span class="st">"Inferred.Menopausal.State"</span>, <span class="st">"Grade"</span>, <span class="st">"Size"</span><span class="op">)</span></span>
<span><span class="va">nFeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/append.html" class="external-link">append</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">scfeatures_result</span>, <span class="kw">function</span><span class="op">(</span><span class="va">metaFeature</span><span class="op">)</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">clinicalAndOmics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/append.html" class="external-link">append</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="va">clinical</span><span class="op">)</span>, <span class="va">scfeatures_result</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### generate classfyr result </span></span>
<span><span class="va">classifyr_result_IMC</span> <span class="op">&lt;-</span> <span class="fu">crossValidate</span><span class="op">(</span><span class="va">clinicalAndOmics</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"timeRFS"</span>, <span class="st">"eventRFS"</span><span class="op">)</span>,</span>
<span>                    extraParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>prepare <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>useFeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="va">usefulFeatures</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                    nFeatures <span class="op">=</span> <span class="va">nFeatures</span>, nFolds <span class="op">=</span> <span class="fl">5</span>, nRepeats <span class="op">=</span> <span class="fl">5</span>, nCores <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">classifyr_result_IMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../../data/classifyr_result_IMC.rds"</span><span class="op">)</span></span></code></pre></div>
<p>Cox proportional hazards is a classical statistical method, as
opposed to machine learning methods like Random survival forest. These
machine learning methods can build remarkably complex relationships
between features, however their running time can be much longer than Cox
proportional hazards. We use feature selection to limit the number of
features considered to at most 100 per metafeature and to save time,
<strong>you can just load the prepared RDS file.</strong> We will
compare the predictive performance between these methods.</p>
<pre><code><span><span class="va">nFeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/append.html" class="external-link">append</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">metaFeature</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">metaFeature</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">survForestCV</span> <span class="op">&lt;-</span> <span class="fu">crossValidate</span><span class="op">(</span><span class="va">clinicalAndOmics</span>, <span class="va">outcome</span>, nFeatures <span class="op">=</span> <span class="va">nFeatures</span>,</span>
<span>                classifier <span class="op">=</span> <span class="st">"randomForest"</span>,</span>
<span>                nFolds <span class="op">=</span> <span class="fl">5</span>, nRepeats <span class="op">=</span> <span class="fl">5</span>, nCores <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">survForestCV</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../../data/survForestCV.RDS"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="tab-pane" id="model-performance-optional" role="tabpanel" aria-labelledby="model-performance-optional-tab">

<p>To examine the distribution of prognostic performance, use
<code>performancePlot</code>. Currently, the only metric for
time-to-event data is C-index and that will automatically be used
because the predictive model type is tracked inside of the result
objects.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Make axis label 45 degree to improve readiability </span></span>
<span><span class="va">tilt</span> <span class="op">&lt;-</span> <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Putting two sets of cross-validation results together</span></span>
<span><span class="va">multiresults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">classifyr_result_IMC</span>, <span class="va">survForestCV</span><span class="op">)</span></span>
<span><span class="va">ordering</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clinical"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">performancePlot</span><span class="op">(</span><span class="va">multiresults</span>,</span>
<span>                characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span>, </span>
<span>                                           row <span class="op">=</span> <span class="st">"Classifier Name"</span><span class="op">)</span>,</span>
<span>                orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="va">ordering</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>                <span class="va">tilt</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<p>Note how the resultant plot is a <code>ggplot2</code> object and can
be further modified. The same code could be used for a categorical
classifier because the random forest implementation provided by the
<code>ranger</code> package has the same interface for both. We will
examine feature selection stability with <code>selectionPlot</code>.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">selectionPlot</span><span class="op">(</span><span class="va">multiresults</span>,</span>
<span>                characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span>, row <span class="op">=</span> <span class="st">"Classifier Name"</span><span class="op">)</span>,</span>
<span>                orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="va">ordering</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="va">tilt</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">distribution</span><span class="op">(</span><span class="va">classifyr_result_IMC</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      assay                   feature proportion</span></span>
<span><span class="co">## 1 clinical Inferred.Menopausal.State          1</span></span></code></pre>
</div>
<div class="tab-pane" id="does-spatial-information-matter" role="tabpanel" aria-labelledby="does-spatial-information-matter-tab">

<p>Does each individual require a different collection of features?
Using <code>samplesMetricMap</code> compare the per-sample C-index for
Cox models for all kinds of metafeatures.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span></span>
<span><span class="fu">samplesMetricMap</span><span class="op">(</span><span class="va">classifyr_result_IMC</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
<pre><code><span><span class="co">## TableGrob (2 x 1) "arrange": 2 grobs</span></span>
<span><span class="co">##   z     cells    name                 grob</span></span>
<span><span class="co">## 1 1 (2-2,1-1) arrange       gtable[layout]</span></span>
<span><span class="co">## 2 2 (1-1,1-1) arrange text[GRID.text.9048]</span></span></code></pre>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="section level3">
<h3 id="explanation-of-spatial-features">Explanation of spatial features<a class="anchor" aria-label="anchor" href="#explanation-of-spatial-features"></a>
</h3>
<ul>
<li>L function:</li>
</ul>
<p>The L function is a spatial statistic used to assess the spatial
distribution of cell-types. It assesses the significance of cell-cell
interactions, by calculating the density of a cell-type with other
cell-types within a certain radius. High values indicate spatial
association (co-localisation), low values indicate spatial
avoidance.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tableau_palette</span> <span class="op">&lt;-</span> <span class="fu">scale_colour_tableau</span><span class="op">(</span> palette <span class="op">=</span> <span class="st">"Tableau 20"</span><span class="op">)</span></span>
<span><span class="va">color_codes</span> <span class="op">&lt;-</span> <span class="va">tableau_palette</span><span class="op">$</span><span class="fu">palette</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co"># select one patient </span></span>
<span><span class="va">one_sample</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span> , <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0128"</span>  <span class="op">]</span></span>
<span><span class="va">one_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu">colData</span><span class="op">(</span><span class="va">one_sample</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">one_sample</span><span class="op">$</span><span class="va">description</span></span>
<span></span>
<span><span class="co"># select certain cell types to examine the interaction </span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span>  <span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span>  <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cells"</span>, <span class="st">"Fibroblasts"</span><span class="op">)</span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span><span class="op">!</span><span class="va">index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"others"</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span><span class="fu">ggplot</span><span class="op">(</span> <span class="va">one_sample</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">celltype</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_codes</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span> <span class="st">"Patient MB-0128 - high L value with \n B cells interacting Fibroblasts"</span><span class="op">)</span></span>
<span> </span>
<span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">one_sample</span><span class="op">$</span><span class="va">description</span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span>  <span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span>  <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Fibroblasts"</span>, <span class="st">"HR- CK7-"</span><span class="op">)</span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span><span class="op">!</span><span class="va">index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"others"</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span> <span class="va">one_sample</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">celltype</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_codes</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span> <span class="st">"Patient MB-0128 - low L value with \n B cells interacting HR_ CK7"</span><span class="op">)</span></span>
<span> </span>
<span><span class="fu">ggarrange</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-32-1.png" width="960"></p>
<ul>
<li>Cell type interaction composition:</li>
</ul>
<p>We calculate the nearest neighbours of each cell and then calculate
the pairs of cell-types based on the nearest neighbour. This allows us
to summarise it into a cell-type interaction composition.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">one_sample</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span> , <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0263"</span>  <span class="op">]</span></span>
<span><span class="va">one_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu">colData</span><span class="op">(</span><span class="va">one_sample</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">one_sample</span><span class="op">$</span><span class="va">description</span></span>
<span> </span>
<span><span class="va">a</span> <span class="op">&lt;-</span><span class="fu">ggplot</span><span class="op">(</span> <span class="va">one_sample</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">celltype</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_codes</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Patient MB-0263"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">feature</span>  <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">celltype_interaction</span></span>
<span><span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span> <span class="va">feature</span><span class="op">[</span> <span class="st">"MB-0263_cond_poor"</span> , <span class="op">]</span><span class="op">)</span>  <span class="op">)</span></span>
<span><span class="va">to_plot</span><span class="op">$</span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"value"</span>, <span class="st">"celltype interaction composition"</span><span class="op">)</span></span>
<span> </span>
<span><span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="va">to_plot</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">$</span><span class="va">value</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="va">to_plot</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span></span>
<span><span class="va">to_plot</span><span class="op">$</span><span class="va">`celltype interaction composition`</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">$</span><span class="va">`celltype interaction composition`</span>, levels <span class="op">=</span> <span class="va">to_plot</span><span class="op">$</span><span class="va">`celltype interaction composition`</span><span class="op">)</span></span>
<span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">to_plot</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span>  <span class="va">`celltype interaction composition`</span>  ,  y <span class="op">=</span> <span class="va">value</span>, fill<span class="op">=</span><span class="va">`celltype interaction composition`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_bar</span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span> <span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"Major cell-type interactions"</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu">ggarrange</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-33-1.png" width="960"></p>
<ul>
<li>Moran’s I:</li>
</ul>
<p>Moran’s I is a spatial autocorrelation statistic based on both
location and values. It quantifies whether similar values tend to occur
near each other or are dispersed.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">high</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span><span class="st">"Ki67"</span>, <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0132"</span>  <span class="op">]</span></span>
<span><span class="va">high_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu">colData</span><span class="op">(</span><span class="va">high</span><span class="op">)</span> <span class="op">)</span> </span>
<span><span class="va">high_meta</span><span class="op">$</span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu">logcounts</span><span class="op">(</span> <span class="va">high</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">low</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span><span class="st">"Ki67"</span>,  <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0249"</span> <span class="op">]</span></span>
<span><span class="va">low_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu">colData</span><span class="op">(</span><span class="va">low</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">low_meta</span><span class="op">$</span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu">logcounts</span><span class="op">(</span><span class="va">low</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">high_meta</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_colour_viridis_c</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Patient MB-0132 - high Moran's I in Ki67"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">low_meta</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_colour_viridis_c</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Patient MB-0249 - low Moran's I in Ki67"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggarrange</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-34-1.png" width="960"></p>
<ul>
<li>Nearest Neighbor Correlation:</li>
</ul>
<p>This metric measures the correlation of proteins/genes between a cell
and its nearest neighbour cell.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">plot_nncorrelation</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">thissample</span> , <span class="va">thisprotein</span><span class="op">)</span><span class="op">{</span></span>
<span>   </span>
<span>       <span class="va">sample_name</span> <span class="op">&lt;-</span> <span class="va">thissample</span></span>
<span>       <span class="va">thissample</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span>, <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span>     <span class="va">sample_name</span><span class="op">]</span></span>
<span>    </span>
<span>      </span>
<span>      <span class="va">exprsMat</span> <span class="op">&lt;-</span> <span class="fu">logcounts</span><span class="op">(</span><span class="va">thissample</span><span class="op">)</span></span>
<span>     </span>
<span>    <span class="co"># calculate NN correlation </span></span>
<span>    <span class="va">cell_points_cts</span> <span class="op">&lt;-</span> <span class="fu">spatstat.geom</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spatstat.geom/man/ppp.html" class="external-link">ppp</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_X</span> <span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span>,</span>
<span>            check <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            xrange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_X</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_X</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="op">)</span>,</span>
<span>            yrange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="op">)</span>,</span>
<span>            marks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">exprsMat</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>    </span>
<span>     <span class="va">value</span> <span class="op">&lt;-</span>  <span class="fu">spatstat.explore</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spatstat.explore/man/nncorr.html" class="external-link">nncorr</a></span><span class="op">(</span><span class="va">cell_points_cts</span><span class="op">)</span><span class="op">[</span><span class="st">"correlation"</span>, <span class="op">]</span></span>
<span>      <span class="va">value</span> <span class="op">&lt;-</span>  <span class="va">value</span><span class="op">[</span>  <span class="va">thisprotein</span><span class="op">]</span></span>
<span>     </span>
<span>    <span class="co"># Find the indices of the two nearest neighbors for each cell</span></span>
<span>    <span class="va">nn_indices</span> <span class="op">&lt;-</span> <span class="fu">nnwhich</span><span class="op">(</span><span class="va">cell_points_cts</span>, k <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">protein</span> <span class="op">&lt;-</span>  <span class="va">thisprotein</span></span>
<span>    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>thiscell_exprs  <span class="op">=</span> <span class="va">exprsMat</span><span class="op">[</span><span class="va">protein</span>, <span class="op">]</span> , exprs <span class="op">=</span>  <span class="va">exprsMat</span><span class="op">[</span><span class="va">protein</span>,<span class="va">nn_indices</span> <span class="op">]</span><span class="op">)</span></span>
<span>    </span>
<span>   <span class="va">p</span> <span class="op">&lt;-</span>  <span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span>, <span class="fu">aes</span><span class="op">(</span> x <span class="op">=</span><span class="va">thiscell_exprs</span> ,  y <span class="op">=</span> <span class="va">exprs</span> , colour <span class="op">=</span>  <span class="va">exprs</span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span> <span class="st">"Patient "</span>, <span class="va">sample_name</span> ,  <span class="st">" nn_corr = "</span> ,  <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">value</span>, <span class="fl">2</span><span class="op">)</span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_colour_viridis_c</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">xlab</span><span class="op">(</span><span class="st">"This cell expression"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"Neighbouring cell expression"</span><span class="op">)</span></span>
<span>   </span>
<span>   <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span> <span class="op">(</span><span class="va">p</span> <span class="op">)</span> </span>
<span></span>
<span><span class="op">}</span></span>
<span></span>
<span>    </span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">plot_nncorrelation</span><span class="op">(</span><span class="st">"MB-0605"</span>,  <span class="st">"HER2"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">plot_nncorrelation</span><span class="op">(</span><span class="st">"MB-0258"</span>,  <span class="st">"HER2"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggarrange</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="CUHK_workshop_v3_files/figure-html/unnamed-chunk-35-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h3>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Debian GNU/Linux 12 (bookworm)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.21.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Australia/Sydney</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] purrr_1.0.2                 simpleSeg_1.2.0            </span></span>
<span><span class="co">##  [3] naniar_1.1.0                reshape_0.8.9              </span></span>
<span><span class="co">##  [5] scran_1.32.0                scater_1.32.0              </span></span>
<span><span class="co">##  [7] scuttle_1.14.0              spatstat_3.1-1             </span></span>
<span><span class="co">##  [9] spatstat.linnet_3.2-1       spatstat.model_3.3-1       </span></span>
<span><span class="co">## [11] rpart_4.1.23                spatstat.explore_3.3-2     </span></span>
<span><span class="co">## [13] nlme_3.1-165                spatstat.random_3.3-1      </span></span>
<span><span class="co">## [15] spatstat.geom_3.3-2         spatstat.univar_3.0-0      </span></span>
<span><span class="co">## [17] spatstat.data_3.1-2         survminer_0.4.9            </span></span>
<span><span class="co">## [19] ggpubr_0.6.0                tidyr_1.3.1                </span></span>
<span><span class="co">## [21] scattermore_1.2             plotly_4.10.4              </span></span>
<span><span class="co">## [23] limma_3.60.6                dplyr_1.1.4                </span></span>
<span><span class="co">## [25] spicyR_1.19.2               ggthemes_5.1.0             </span></span>
<span><span class="co">## [27] lisaClust_1.12.3            ClassifyR_3.9.1            </span></span>
<span><span class="co">## [29] survival_3.6-4              BiocParallel_1.38.0        </span></span>
<span><span class="co">## [31] MultiAssayExperiment_1.30.0 generics_0.1.3             </span></span>
<span><span class="co">## [33] scFeatures_1.3.4            ggplot2_3.5.1              </span></span>
<span><span class="co">## [35] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0</span></span>
<span><span class="co">## [37] Biobase_2.64.0              GenomicRanges_1.56.0       </span></span>
<span><span class="co">## [39] GenomeInfoDb_1.40.1         IRanges_2.38.0             </span></span>
<span><span class="co">## [41] S4Vectors_0.42.0            BiocGenerics_0.50.0        </span></span>
<span><span class="co">## [43] MatrixGenerics_1.16.0       matrixStats_1.3.0          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] SpatialExperiment_1.14.0   R.methodsS3_1.8.2         </span></span>
<span><span class="co">##   [3] GSEABase_1.66.0            tiff_0.1-12               </span></span>
<span><span class="co">##   [5] EnsDb.Mmusculus.v79_2.99.0 goftest_1.2-3             </span></span>
<span><span class="co">##   [7] Biostrings_2.72.0          HDF5Array_1.32.0          </span></span>
<span><span class="co">##   [9] vctrs_0.6.5                digest_0.6.37             </span></span>
<span><span class="co">##  [11] png_0.1-8                  shape_1.4.6.1             </span></span>
<span><span class="co">##  [13] EnsDb.Hsapiens.v79_2.99.0  ggrepel_0.9.5             </span></span>
<span><span class="co">##  [15] deldir_2.0-4               magick_2.8.3              </span></span>
<span><span class="co">##  [17] MASS_7.3-60.2              pkgdown_2.0.9             </span></span>
<span><span class="co">##  [19] reshape2_1.4.4             foreach_1.5.2             </span></span>
<span><span class="co">##  [21] httpuv_1.6.15              withr_3.0.0               </span></span>
<span><span class="co">##  [23] xfun_0.43                  memoise_2.0.1             </span></span>
<span><span class="co">##  [25] proxyC_0.4.1               cytomapper_1.16.0         </span></span>
<span><span class="co">##  [27] ggbeeswarm_0.7.2           systemfonts_1.0.6         </span></span>
<span><span class="co">##  [29] ragg_1.3.0                 zoo_1.8-12                </span></span>
<span><span class="co">##  [31] GlobalOptions_0.1.2        gtools_3.9.5              </span></span>
<span><span class="co">##  [33] V8_4.4.2                   SingleCellSignalR_1.16.0  </span></span>
<span><span class="co">##  [35] R.oo_1.26.0                KEGGREST_1.44.0           </span></span>
<span><span class="co">##  [37] promises_1.3.0             httr_1.4.7                </span></span>
<span><span class="co">##  [39] rstatix_0.7.2              restfulr_0.0.15           </span></span>
<span><span class="co">##  [41] rhdf5filters_1.16.0        rhdf5_2.48.0              </span></span>
<span><span class="co">##  [43] rstudioapi_0.16.0          UCSC.utils_1.0.0          </span></span>
<span><span class="co">##  [45] concaveman_1.1.0           babelgene_22.9            </span></span>
<span><span class="co">##  [47] curl_6.2.2                 zlibbioc_1.50.0           </span></span>
<span><span class="co">##  [49] ScaledMatrix_1.12.0        polyclip_1.10-6           </span></span>
<span><span class="co">##  [51] GenomeInfoDbData_1.2.12    SparseArray_1.4.0         </span></span>
<span><span class="co">##  [53] fftwtools_0.9-11           xtable_1.8-4              </span></span>
<span><span class="co">##  [55] stringr_1.5.1              desc_1.4.3                </span></span>
<span><span class="co">##  [57] evaluate_1.0.3             S4Arrays_1.4.1            </span></span>
<span><span class="co">##  [59] irlba_2.3.5.1              colorspace_2.1-1          </span></span>
<span><span class="co">##  [61] magrittr_2.0.3             later_1.3.2               </span></span>
<span><span class="co">##  [63] viridis_0.6.5              lattice_0.22-7            </span></span>
<span><span class="co">##  [65] cowplot_1.1.3              XML_3.99-0.16.1           </span></span>
<span><span class="co">##  [67] ggupset_0.3.0              svgPanZoom_0.3.4          </span></span>
<span><span class="co">##  [69] class_7.3-22               pillar_1.9.0              </span></span>
<span><span class="co">##  [71] iterators_1.0.14           EBImage_4.46.0            </span></span>
<span><span class="co">##  [73] caTools_1.18.2             compiler_4.4.1            </span></span>
<span><span class="co">##  [75] beachmat_2.20.0            stringi_1.8.3             </span></span>
<span><span class="co">##  [77] tensor_1.5                 minqa_1.2.7               </span></span>
<span><span class="co">##  [79] GenomicAlignments_1.40.0   plyr_1.8.9                </span></span>
<span><span class="co">##  [81] msigdbr_7.5.1              crayon_1.5.3              </span></span>
<span><span class="co">##  [83] abind_1.4-8                BiocIO_1.14.0             </span></span>
<span><span class="co">##  [85] sp_2.1-4                   locfit_1.5-9.9            </span></span>
<span><span class="co">##  [87] terra_1.7-71               bit_4.6.0                 </span></span>
<span><span class="co">##  [89] codetools_0.2-20           textshaping_0.3.7         </span></span>
<span><span class="co">##  [91] BiocSingular_1.20.0        bslib_0.7.0               </span></span>
<span><span class="co">##  [93] multtest_2.60.0            mime_0.12                 </span></span>
<span><span class="co">##  [95] splines_4.4.1              circlize_0.4.16           </span></span>
<span><span class="co">##  [97] Rcpp_1.0.12                sparseMatrixStats_1.16.0  </span></span>
<span><span class="co">##  [99] knitr_1.46                 blob_1.2.4                </span></span>
<span><span class="co">## [101] utf8_1.2.4                 AnnotationFilter_1.28.0   </span></span>
<span><span class="co">## [103] lme4_1.1-35.3              fs_1.6.6                  </span></span>
<span><span class="co">## [105] nnls_1.5                   DelayedMatrixStats_1.26.0 </span></span>
<span><span class="co">## [107] GSVA_1.52.0                ggsignif_0.6.4            </span></span>
<span><span class="co">## [109] tibble_3.2.1               Matrix_1.7-0              </span></span>
<span><span class="co">## [111] scam_1.2-16                statmod_1.5.0             </span></span>
<span><span class="co">## [113] svglite_2.1.3              visdat_0.6.0              </span></span>
<span><span class="co">## [115] tweenr_2.0.3               pkgconfig_2.0.3           </span></span>
<span><span class="co">## [117] pheatmap_1.0.12            tools_4.4.1               </span></span>
<span><span class="co">## [119] cachem_1.0.8               RSQLite_2.3.6             </span></span>
<span><span class="co">## [121] viridisLite_0.4.2          DBI_1.2.3                 </span></span>
<span><span class="co">## [123] numDeriv_2016.8-1.1        fastmap_1.2.0             </span></span>
<span><span class="co">## [125] rmarkdown_2.26             scales_1.3.0              </span></span>
<span><span class="co">## [127] shinydashboard_0.7.2       Rsamtools_2.20.0          </span></span>
<span><span class="co">## [129] broom_1.0.5                sass_0.4.9                </span></span>
<span><span class="co">## [131] BiocManager_1.30.25        graph_1.82.0              </span></span>
<span><span class="co">## [133] carData_3.0-5              farver_2.1.2              </span></span>
<span><span class="co">## [135] mgcv_1.9-1                 yaml_2.3.8                </span></span>
<span><span class="co">## [137] rtracklayer_1.64.0         cli_3.6.5                 </span></span>
<span><span class="co">## [139] lifecycle_1.0.4            bluster_1.14.0            </span></span>
<span><span class="co">## [141] backports_1.5.0            annotate_1.82.0           </span></span>
<span><span class="co">## [143] gtable_0.3.5               rjson_0.2.21              </span></span>
<span><span class="co">## [145] parallel_4.4.1             ape_5.8                   </span></span>
<span><span class="co">## [147] jsonlite_2.0.0             edgeR_4.2.0               </span></span>
<span><span class="co">## [149] bitops_1.0-9               bit64_4.0.5               </span></span>
<span><span class="co">## [151] Rtsne_0.17                 spatstat.utils_3.1-0      </span></span>
<span><span class="co">## [153] BiocNeighbors_1.22.0       bdsmatrix_1.3-7           </span></span>
<span><span class="co">## [155] highr_0.10                 jquerylib_0.1.4           </span></span>
<span><span class="co">## [157] metapod_1.12.0             dqrng_0.4.1               </span></span>
<span><span class="co">## [159] survMisc_0.5.6             R.utils_2.12.3            </span></span>
<span><span class="co">## [161] lazyeval_0.2.2             shiny_1.8.1.1             </span></span>
<span><span class="co">## [163] htmltools_0.5.8.1          KMsurv_0.1-5              </span></span>
<span><span class="co">## [165] ensembldb_2.28.0           glue_1.8.0                </span></span>
<span><span class="co">## [167] XVector_0.44.0             RCurl_1.98-1.14           </span></span>
<span><span class="co">## [169] coxme_2.2-22               jpeg_0.1-11               </span></span>
<span><span class="co">## [171] gridExtra_2.3              AUCell_1.26.0             </span></span>
<span><span class="co">## [173] boot_1.3-31                igraph_2.0.3              </span></span>
<span><span class="co">## [175] R6_2.5.1                   gplots_3.1.3.1            </span></span>
<span><span class="co">## [177] labeling_0.4.3             km.ci_0.5-6               </span></span>
<span><span class="co">## [179] ggh4x_0.2.8                GenomicFeatures_1.56.0    </span></span>
<span><span class="co">## [181] cluster_2.1.8.1            Rhdf5lib_1.26.0           </span></span>
<span><span class="co">## [183] nloptr_2.0.3               DelayedArray_0.30.0       </span></span>
<span><span class="co">## [185] tidyselect_1.2.1           vipor_0.4.7               </span></span>
<span><span class="co">## [187] ProtGenerics_1.36.0        raster_3.6-26             </span></span>
<span><span class="co">## [189] ggforce_0.4.2              car_3.1-2                 </span></span>
<span><span class="co">## [191] AnnotationDbi_1.66.0       rsvd_1.0.5                </span></span>
<span><span class="co">## [193] munsell_0.5.1              KernSmooth_2.23-26        </span></span>
<span><span class="co">## [195] data.table_1.17.0          htmlwidgets_1.6.4         </span></span>
<span><span class="co">## [197] RColorBrewer_1.1-3         rlang_1.1.4               </span></span>
<span><span class="co">## [199] spatstat.sparse_3.1-0      lmerTest_3.1-3            </span></span>
<span><span class="co">## [201] ggnewscale_0.4.10          fansi_1.0.6               </span></span>
<span><span class="co">## [203] beeswarm_0.4.0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h3>
<p>The authors thank all their colleagues, particularly at The
University of Sydney, Sydney Precision Data Science and Charles Perkins
Centre for their support and intellectual engagement. Special thanks to
Ellis Patrick, Shila Ghazanfar, Andy Tran, Helen, and Daniel for guiding
and supporting the building of this workshop.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sydney Data Science Centre.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
