<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Unlocking single cell spatial omics analyses with scdney • NUS2025</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="NUS_workshop_v1_files/libs/quarto-html/popper.min.js"></script><script src="NUS_workshop_v1_files/libs/quarto-html/tippy.umd.min.js"></script><link href="NUS_workshop_v1_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="NUS_workshop_v1_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Unlocking single cell spatial omics analyses with scdney">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">NUS2025</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/NUS_workshop_v1.html">Unlocking single cell spatial omics analyses with scdney</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Unlocking single cell spatial omics analyses with scdney</h1>


      <p class="date">2025-06-13</p>


      <div class="d-none name"><code></code></div>
    </div>






    <style>
.question {
  padding: 1em;
  background: lightcyan;
  color: black;
  border-radius: 10px;
}
</style>
<p><strong>Presenting authors</strong></p>
<p>Harry Robertson<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn><mo>,</mo><mn>3</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2,3}</annotation></semantics></math>, Lijia Yu<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2}</annotation></semantics></math>, Beilei Bian<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2}</annotation></semantics></math>, Andrew Zhang<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>4</mn></mrow></msup><annotation encoding="application/x-tex">^{1,4}</annotation></semantics></math>, Jean Yang<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn><mo>,</mo><mn>3</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2,3}</annotation></semantics></math>.</p>
<p><strong>Contributors</strong></p>
<p>Yue Cao<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2}</annotation></semantics></math>, Lijia Yu<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2}</annotation></semantics></math>, Andy Tran<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2}</annotation></semantics></math>, Daniel Kim<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>3</mn></mrow></msup><annotation encoding="application/x-tex">^{1,3}</annotation></semantics></math>, Dario Strbenac<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2}</annotation></semantics></math>, Nicholas Robertson<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mn>1</mn></msup><annotation encoding="application/x-tex">^{1}</annotation></semantics></math>, Helen Fu<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>3</mn><mo>,</mo><mn>4</mn></mrow></msup><annotation encoding="application/x-tex">^{3,4}</annotation></semantics></math>, Jean Yang<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn><mo>,</mo><mn>4</mn></mrow></msup><annotation encoding="application/x-tex">^{1,2,4}</annotation></semantics></math>.</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mn>1</mn></msup><annotation encoding="application/x-tex">^1</annotation></semantics></math> Sydney Precision Data Science Centre, University of Sydney, Australia<br><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mn>2</mn></msup><annotation encoding="application/x-tex">^2</annotation></semantics></math> School of Mathematics and Statistics, University of Sydney, Australia<br><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mn>3</mn></msup><annotation encoding="application/x-tex">^3</annotation></semantics></math> Charles Perkins Centre, University of Sydney, Australia<br><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mn>4</mn></msup><annotation encoding="application/x-tex">^4</annotation></semantics></math> School of Computer Science, University of Sydney, Australia</p>
<p><br> Contact: jean.yang@sydney.edu.au</p>
<section class="section level2"><h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The emergence of high-resolution spatial omics technologies has revolutionized our ability to map cellular ecosystems in situ. This workshop explores approaches in multi-sample spatial data analysis. We cover end-to-end workflow considerations—from experimental design and QC to spatial feature interpretation—with case studies in disease prediction.</p>
<div class="panel-tabset">
<ul id="tabset-1" class="panel-tabset-tabby">
<li><a data-tabby-default="" href="#tabset-1-1">Our dataset</a></li>
<li><a href="#tabset-1-2">Assumed knowledge</a></li>
<li><a href="#tabset-1-3">Learning objectives</a></li>
</ul>
<ul class="nav nav-tabs" id="NA" role="tablist"><li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<p>A subset of the widely-known METABRIC breast cancer cohort has recently been analyzed using imaging mass cytometry: <a href="https://www.nature.com/articles/s43018-020-0026-6" class="external-link">Imaging Mass Cytometry and Multiplatform Genomics Define the Phenogenomic Landscape of Breast Cancer</a>. Patient clinical data was sourced from the Supplementary Table 5 of <a href="https://www.nature.com/articles/s41586-019-1007-8" class="external-link">Dynamics of Breast-cancer Relapse Reveal Late-recurring ER-positive Genomic Subgroups</a>.</p>
</button></li></ul>
<div class="tab-content"><div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab">

<div id="tabset-1-2">
<ul>
<li>Experience with R.</li>
<li>Familiarity with the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment class</a>.</li>
<li>Basic knowledge in <a href="https://bioconductor.org/books/release/OSCA/index.html" class="external-link">single cell data analysis</a>. You can access our <a href="https://github.com/SydneyBioX/scdney#scdney-workshops-series" class="external-link">previous workshops</a> for a quick tutorial.</li>
<li>Ability to install all required R packages, please check <code>sessionInfo</code> at the end of this document to ensure you are using the correct versions.</li>
<li>Familiarity with our previous workshop vignette on <a href="https://sydneybiox.github.io/BIS2019_SC/" class="external-link">Introduction to Single Cell RNA-seq Analysis</a>.</li>
</ul>
</div>
<div id="tabset-1-3">
<ul>
<li>Describe and visualise spatial omics datasets.</li>
<li>Understand approaches for quality control of spatial proteomics data.</li>
<li>Calculate spatial statistics at the cell-type level using <code>scFeatures</code>.</li>
<li>Perform multi-view disease outcome prediction with the package <code>ClassifyR</code>.</li>
<li>Develop an understanding of:
<ul>
<li>evaluation of classification and survival models .</li>
<li>evaluate cohort heterogeneity given a survival model.</li>
</ul>
</li>
<li>Explore various strategies for disease outcome prognosis using spatial omics data.</li>
</ul>
</div>
</div></div>
</div>
<div class="question">
<p><strong>Our question of interest</strong></p>
<p>We want to know if the risk of recurrence in the METABRIC breast cancer cohort can be accurately estimated to inform how aggressively they need to be treated.</p>
</div>
</section><section class="section level2"><h2 id="part-0-loading-an-example-image-in-r">Part 0: Loading an example image in R<a class="anchor" aria-label="anchor" href="#part-0-loading-an-example-image-in-r"></a>
</h2>
<p>To begin, we will start with loading an example image in R. It is not neccessary for you to be familiar with this part, as we will provide you with pre-processed data for the rest of the workshop.</p>
<div class="panel-tabset">
<ul id="tabset-2" class="panel-tabset-tabby">
<li><a data-tabby-default="" href="#tabset-2-1">DNA staining image</a></li>
<li><a href="#tabset-2-2">Mask</a></li>
<li><a href="#tabset-2-3">Overlay</a></li>
</ul>
<ul class="nav nav-tabs" id="NA" role="tablist"><li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BodenmillerGroup/cytomapper" class="external-link">cytomapper</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aoles/EBImage" class="external-link">EBImage</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/S4Vectors" class="external-link">S4Vectors</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">img_path</span>  <span class="op">&lt;-</span> <span class="st">"data/raw_image_example/MB0002_1_345_fullstack.tiff"</span></span>
<span><span class="va">mask_path</span> <span class="op">&lt;-</span> <span class="st">"data/raw_image_example/MB0002_1_345_cellmask.tiff"</span></span>
<span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="st">"data/raw_image_example/MB0002_1_345"</span></span>
<span></span>
<span><span class="co">#Load image (multi-channel)</span></span>
<span><span class="va">images</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/loadImages.html" class="external-link">loadImages</a></span><span class="op">(</span><span class="va">img_path</span>, single_channel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_id</span></span>
<span><span class="co">#Load mask</span></span>
<span><span class="va">mask_img</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/io.html" class="external-link">readImage</a></span><span class="op">(</span><span class="va">mask_path</span>,as.is <span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">masks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/CytoImageList.html" class="external-link">CytoImageList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">mask_img</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">masks</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_id</span></span>
<span></span>
<span><span class="co"># Add metadata rows so img_id can be matched by COLUMN NAME</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">masks</span><span class="op">)</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">masks</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Build channel names from CSVs and assign them</span></span>
<span><span class="va">metal_order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"data/raw_image_example/channel_to_metal_order.csv"</span>,</span>
<span>                        header <span class="op">=</span> <span class="cn">FALSE</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">metal_order</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"metal"</span></span>
<span><span class="va">metal_order</span><span class="op">$</span><span class="va">Channel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">metal_order</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metal_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"data/raw_image_example/metal_to_marker.csv"</span>,</span>
<span>                      header <span class="op">=</span> <span class="cn">TRUE</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">metal_order</span>, <span class="va">metal_map</span>, by <span class="op">=</span> <span class="st">"metal"</span>, all.x <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="va">merged</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">merged</span><span class="op">$</span><span class="va">Channel</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">merged</span><span class="op">$</span><span class="va">marker</span><span class="op">)</span>, <span class="va">merged</span><span class="op">$</span><span class="va">metal</span>, <span class="va">merged</span><span class="op">$</span><span class="va">marker</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign channel names to the multi-channel image</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/CytoImageList-naming.html" class="external-link">channelNames</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">labels</span></span>
<span><span class="va">marker</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html" class="external-link">plotPixels</a></span><span class="op">(</span></span>
<span>  image <span class="op">=</span> <span class="va">images</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"Total HH3"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-2-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</button></li></ul>
<div class="tab-content"><div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab">

<div id="tabset-2-2">
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the mask</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotCells.html" class="external-link">plotCells</a></span><span class="op">(</span><span class="va">masks</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-3-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-2-3">
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot overlay </span></span>
<span><span class="va">combine</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html" class="external-link">plotPixels</a></span><span class="op">(</span></span>
<span>  image    <span class="op">=</span> <span class="va">images</span>,</span>
<span>  mask     <span class="op">=</span> <span class="va">masks</span>,</span>
<span>  img_id   <span class="op">=</span> <span class="st">"sample_id"</span>,      <span class="co"># column name in mcols(images)/mcols(masks)</span></span>
<span>  colour_by <span class="op">=</span> <span class="st">"Total HH3"</span>     </span>
<span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-4-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div>
</div></div>
</div>
</section><section class="section level2"><h2 id="part-1-exploring-the-data">Part 1: Exploring the data<a class="anchor" aria-label="anchor" href="#part-1-exploring-the-data"></a>
</h2>
<section class="level2"><h3 id="initial-data-exploration">1.1: Initial data exploration<a class="anchor" aria-label="anchor" href="#initial-data-exploration"></a>
</h3>
<p>At the start of any analysis pipeline, it is often good to explore the data to get a sense of the structure and its complexity. Let’s explore the data to answer the questions below:</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol type="1">
<li>How many features and observations are there in the data and what do the features represent?</li>
<li>What covariates are in our data?</li>
<li>Given our question of interest, what variable would be our outcome variable?</li>
</ol>
</div>
<p><br></p>
<div class="panel-tabset">
<ul id="tabset-3" class="panel-tabset-tabby">
<li><a data-tabby-default="" href="#tabset-3-1">SCE object</a></li>
<li><a href="#tabset-3-2">Data overview</a></li>
<li><a href="#tabset-3-3">Covariates</a></li>
</ul>
<ul class="nav nav-tabs" id="NA" role="tablist"><li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<p>First, we take a quick look at the structure of the SingleCell Experiment object.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"data/breastCancer.RData"</span><span class="op">)</span></span>
<span><span class="va">data_sce</span> <span class="op">=</span> <span class="va">IMC</span></span>
<span></span>
<span><span class="co"># Structure of our data</span></span>
<span><span class="va">data_sce</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment
dim: 38 76307
metadata(0):
assays(2): counts logcounts
rownames(38): HH3_total CK19 ... H3K27me3 CK5
rowData names(0):
colnames(76307): MB-0002:345:93 MB-0002:345:107 ... MB-0663:394:577
  MB-0663:394:578
colData names(17): file_id metabricId ... x_cord y_cord
reducedDimNames(1): UMAP
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Metadata</span></span>
<span><span class="co">#DT::datatable(data.frame(colData(data_sce)))</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>                     file_id metabricId core_id ImageNumber ObjectNumber
MB-0002:345:93  MB0002_1_345    MB-0002       1         345           93
MB-0002:345:107 MB0002_1_345    MB-0002       1         345          107
MB-0002:345:113 MB0002_1_345    MB-0002       1         345          113
MB-0002:345:114 MB0002_1_345    MB-0002       1         345          114
MB-0002:345:125 MB0002_1_345    MB-0002       1         345          125
MB-0002:345:127 MB0002_1_345    MB-0002       1         345          127
                Fibronectin Location_Center_X Location_Center_Y SOM_nodes
MB-0002:345:93    0.4590055          179.7088          27.90110       130
MB-0002:345:107   0.1176471          193.0588          29.29412       131
MB-0002:345:113   0.2512903          162.1935          28.83871        83
MB-0002:345:114   0.4788444          198.3111          29.15556       128
MB-0002:345:125   6.1901112          262.8889          31.26667         7
MB-0002:345:127   0.6897432          136.0946          32.98649       130
                pg_cluster    description   region      ki67
MB-0002:345:93          54       HR- CK7- region_1  low_ki67
MB-0002:345:107         54       HR- CK7- region_4  low_ki67
MB-0002:345:113         28    HRlow CKlow region_1 high_ki67
MB-0002:345:114         54       HR- CK7- region_4 high_ki67
MB-0002:345:125         11 Myofibroblasts region_2  low_ki67
MB-0002:345:127         54       HR- CK7- region_1  low_ki67
                              celltype  sample   x_cord   y_cord
MB-0002:345:93        HR- CK7--region1 MB-0002 179.7088 27.90110
MB-0002:345:107       HR- CK7--region4 MB-0002 193.0588 29.29412
MB-0002:345:113    HRlow CKlow-region1 MB-0002 162.1935 28.83871
MB-0002:345:114       HR- CK7--region4 MB-0002 198.3111 29.15556
MB-0002:345:125 Myofibroblasts-region2 MB-0002 262.8889 31.26667
MB-0002:345:127       HR- CK7--region1 MB-0002 136.0946 32.98649</code></pre>
</div>
</div>
</button></li></ul>
<div class="tab-content"><div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab">

<div id="tabset-3-2">
<p>Here, we take a quick look at what our rows and columns represent and the dimensions of our data.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Glimpse the first few observations </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>[1] "MB-0002:345:93"  "MB-0002:345:107" "MB-0002:345:113" "MB-0002:345:114"
[5] "MB-0002:345:125" "MB-0002:345:127"</code></pre>
</div>
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Glimpse the first few features</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>[1] "HH3_total" "CK19"      "CK8_18"    "Twist"     "CD68"      "CK14"     </code></pre>
</div>
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># How many features and observations are in our dataset?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>[1]    38 76307</code></pre>
</div>
</div>
</div>
<div id="tabset-3-3">
<p>In addition to the proteomics data, it’s important to understand what other covariates, such as clinical variables, are in our dataset. These can help us in answering our question(s) of interest or formulate new questions. For example, we can’t explore the association between smoking status and breast cancer severity if the variable for smoking status doesn’t exist in our data.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Explore covariates</span></span>
<span></span>
<span><span class="co"># DT::datatable(clinical)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code> [1] "file_id"           "metabricId"        "core_id"
 [4] "ImageNumber"       "ObjectNumber"      "Fibronectin"
 [7] "Location_Center_X" "Location_Center_Y" "SOM_nodes"
[10] "pg_cluster"        "description"       "region"
[13] "ki67"              "celltype"          "sample"
[16] "x_cord"            "y_cord"           </code></pre>
</div>
</div>
</div>
</div></div>
</div>
</section><section class="level2"><h3 id="is-this-a-complex-dataset">1.2: Is this a complex dataset?<a class="anchor" aria-label="anchor" href="#is-this-a-complex-dataset"></a>
</h3>
<p>Now that we have a basic idea of what our data looks like, we can look at it in more detail. While initial data exploration reveals fundamental patterns, deeper examination is very helpful. As it serves two critical purposes: first, to detect anomalies or biases requiring remediation, and second, to inform our choice of analytical methods tailored to the biological questions at hand.</p>
<div class="panel-tabset">
<ul id="tabset-4" class="panel-tabset-tabby">
<li><a data-tabby-default="" href="#tabset-4-1">Missing values</a></li>
<li><a href="#tabset-4-2">Imbalance</a></li>
<li><a href="#tabset-4-3">Relationships</a></li>
<li><a href="#tabset-4-4">Visualise cells</a></li>
</ul>
<ul class="nav nav-tabs" id="NA" role="tablist"><li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<p>Examining the proportion of missing values in a dataset is crucial for ensuring the accuracy and validity of data analysis. Missing values can significantly impact the reliability of statistical results, potentially leading to biased conclusions and reduced statistical power. Here, we plot the proportion of missing values for each of our clinical variables.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># What is the proportion of missing values in our clinical data?</span></span>
<span><span class="fu">gg_miss_var</span><span class="op">(</span><span class="va">clinical</span>, show_pct <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-7-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</button></li></ul>
<div class="tab-content"><div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab">

<div id="tabset-4-2">
<p>Looking at imbalance in data is crucial because it can lead to biased models and inaccurate predictions, especially in classification tasks. Imbalance occurs when one class is significantly underrepresented compared to others. This can cause models to be overly influenced by the majority class, leading to poor performance on the minority class. Here, we tabularise some of the relevant clinical variables and plot the distribution of the <code>Time to Recurrence-Free Survival</code> variable.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estrogen receptor status</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">clinical</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">ER.Status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_bar</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#0099B4"</span>, color <span class="op">=</span> <span class="st">"white"</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Count\n"</span>, </span>
<span>       x <span class="op">=</span> <span class="st">"\nER status"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
</figure>
</div>
</div>
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Histological type</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">clinical</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Histological.Type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_bar</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#fc6203"</span>, color <span class="op">=</span> <span class="st">"white"</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Count\n"</span>, </span>
<span>       x <span class="op">=</span> <span class="st">"\nHistological Type"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-8-2.png" width="672"></p>
</figure>
</div>
</div>
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># eventRFS: "Event in Recurrence-Free Survival."It indicates whether the event has occurred.#</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">clinical</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">eventRFS</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_bar</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#40dbb2"</span>, color <span class="op">=</span> <span class="st">"white"</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Count\n"</span>, </span>
<span>       x <span class="op">=</span> <span class="st">"\nEvent in Recurrence-Free Survival"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-8-3.png" width="672"></p>
</figure>
</div>
</div>
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># timeRFS: "Time to Recurrence-Free Survival." It is the time period until recurrence occurs. </span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">clinical</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">timeRFS</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#de9921"</span>, color <span class="op">=</span> <span class="st">"white"</span>, alpha <span class="op">=</span> <span class="fl">0.8</span>, bins<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Frequency\n"</span>, </span>
<span>       x <span class="op">=</span> <span class="st">"\ntimeRFS"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-8-4.png" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-4-3">
<p>Here, we explore the distribution of the outcomes and variables in the meta-data. We use cross-tabulation to examine the following variables: Surgery vs death, ER status, and Grade.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Stage and death</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">$</span><span class="va">Breast.Surgery</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">Death</span>, useNA <span class="op">=</span> <span class="st">"ifany"</span><span class="op">)</span> </span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>
                     0  1
  BREAST CONSERVING 38 14
  MASTECTOMY        16  6
  &lt;NA&gt;               2  1</code></pre>
</div>
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ER status and grade</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">$</span><span class="va">ER.Status</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">Grade</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>
       1  2  3
  neg  0  0 11
  pos 13 30 17</code></pre>
</div>
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># "Number of individuals based on Grade</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">$</span><span class="va">Grade</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">Death</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>
     0  1
  1 12  2
  2 25  5
  3 17 11</code></pre>
</div>
</div>
</div>
<div id="tabset-4-4">
<p>To assess potential batch effects and sample-specific clustering patterns, we visualize the cells in our data colored by the sample of origin. This qualitative inspection provides an intuitive first assessment of any batch effects in our data integration. If there is batch effects then we’d need to apply batch correction to resolve this issue.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#data_sce &lt;- runUMAP(data_sce, scale=TRUE)</span></span>
<span></span>
<span><span class="co"># With the UMAP function we can highlight by meta data of interest</span></span>
<span><span class="co"># Here we highlight the UMAP by sample ID</span></span>
<span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">data_sce</span>, colour_by <span class="op">=</span> <span class="st">"metabricId"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">coord_equal</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-10-1.png" width="1152"></p>
</figure>
</div>
</div>
</div>
</div>
</div></div>
</div>
<div class="question">
<p><em>Discussion</em></p>
<ul>
<li>Are there any anomalies or biases that might need to be corrected to ensure our analysis is robust?</li>
<li>Given our question of interest and the characteristics of our data, are there any particular analytic techniques that would be appropriate? Are there any that would not be appropriate?</li>
</ul>
</div>
<p><br></p>
</section></section><section class="section level2"><h2 id="part-2-quality-control">Part 2: Quality control<a class="anchor" aria-label="anchor" href="#part-2-quality-control"></a>
</h2>
<p>Here, we explore key approaches for evaluating the quality of IMC data. A robust assessment should consider multiple factors, including the density of marker expression, marker correlations, and their co-expression. Something we might want to think about is what characeristics might indicate whether a sample(s) is low/high quality?</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_counts</span> <span class="op">&lt;-</span> <span class="va">IMC</span><span class="op">@</span><span class="va">colData</span> <span class="op">|&gt;</span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">metabricId</span>, name <span class="op">=</span> <span class="st">"cell_count"</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="co"># Count rows per metabricId</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">cell_count</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">IMC</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">IMC</span><span class="op">@</span><span class="va">colData</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Location_Center_X"</span>,<span class="st">"Location_Center_Y"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="va">cell_type_mapping</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"B cells"</span> <span class="op">=</span> <span class="st">"B cell"</span>,</span>
<span>  <span class="st">"T cells"</span> <span class="op">=</span> <span class="st">"T cell"</span>,</span>
<span>  <span class="st">"Macrophages Vim+ CD45low"</span> <span class="op">=</span> <span class="st">"Macrophage"</span>,</span>
<span>  <span class="st">"Macrophages Vim+ Slug-"</span> <span class="op">=</span> <span class="st">"Macrophage"</span>,</span>
<span>  <span class="st">"Macrophages Vim+ Slug+"</span> <span class="op">=</span> <span class="st">"Macrophage"</span>,</span>
<span></span>
<span>  <span class="st">"Endothelial"</span> <span class="op">=</span> <span class="st">"Endothelial"</span>,</span>
<span></span>
<span>  <span class="st">"Fibroblasts"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span>  <span class="st">"Fibroblasts CD68+"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span>  <span class="st">"Myofibroblasts"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span>  <span class="st">"Vascular SMA+"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span></span>
<span>  <span class="st">"Myoepithelial"</span> <span class="op">=</span> <span class="st">"Myoepithelial"</span>,</span>
<span></span>
<span>  <span class="st">"HR+ CK7-"</span> <span class="op">=</span> <span class="st">"Tumor HR+"</span>,</span>
<span>  <span class="st">"HR+ CK7- Ki67+"</span> <span class="op">=</span> <span class="st">"Tumor HR+"</span>,</span>
<span>  <span class="st">"HR+ CK7- Slug+"</span> <span class="op">=</span> <span class="st">"Tumor HR+"</span>,</span>
<span></span>
<span>  <span class="st">"HR- CK7+"</span> <span class="op">=</span> <span class="st">"Tumor HR-"</span>,</span>
<span>  <span class="st">"HR- CK7-"</span> <span class="op">=</span> <span class="st">"Tumor HR-"</span>,</span>
<span>  <span class="st">"HR- Ki67+"</span> <span class="op">=</span> <span class="st">"Tumor HR-"</span>,</span>
<span>  <span class="st">"HR- CKlow CK5+"</span> <span class="op">=</span> <span class="st">"Tumor HR-"</span>,</span>
<span></span>
<span>  <span class="st">"HRlow CKlow"</span> <span class="op">=</span> <span class="st">"Tumor HR-low/mixed"</span>,</span>
<span>  <span class="st">"HER2+"</span> <span class="op">=</span> <span class="st">"Tumor HER2+"</span>,</span>
<span>  <span class="st">"Basal CKlow"</span> <span class="op">=</span> <span class="st">"Tumor Basal-like"</span>,</span>
<span>  <span class="st">"Hypoxia"</span> <span class="op">=</span> <span class="st">"Tumor Hypoxic"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert IMC description to higher-level categories</span></span>
<span><span class="va">IMC</span><span class="op">$</span><span class="va">high_level_category</span> <span class="op">&lt;-</span> <span class="fu">recode</span><span class="op">(</span><span class="va">IMC</span><span class="op">$</span><span class="va">description</span>, <span class="op">!</span><span class="op">!</span><span class="op">!</span><span class="va">cell_type_mapping</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">imc.spe</span><span class="op">=</span><span class="va">IMC</span><span class="op">[</span>,<span class="va">IMC</span><span class="op">$</span><span class="va">metabricId</span><span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span><span class="va">cell_counts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,<span class="op">]</span><span class="op">$</span><span class="va">metabricId</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Convert IMC description to higher-level categories</span></span>
<span><span class="va">imc.spe</span><span class="op">$</span><span class="va">high_level_category</span> <span class="op">&lt;-</span> <span class="fu">recode</span><span class="op">(</span><span class="va">imc.spe</span><span class="op">$</span><span class="va">description</span>, <span class="op">!</span><span class="op">!</span><span class="op">!</span><span class="va">cell_type_mapping</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu">plot_marker_densities</span><span class="op">(</span><span class="va">imc.spe</span>, sample_id_col <span class="op">=</span> <span class="st">"metabricId"</span>,</span>
<span>                                markers_to_plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD20"</span>,<span class="st">"CD3"</span>,<span class="st">"CD68"</span>,<span class="st">"vWF_CD31"</span>,<span class="st">"Vimentin"</span>,<span class="st">"HER2"</span><span class="op">)</span>,</span>
<span>                               assay_name <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<section class="level2"><h3 id="qc-for-the-panel-markers">2.1: QC for the panel / markers<a class="anchor" aria-label="anchor" href="#qc-for-the-panel-markers"></a>
</h3>
<p>Here, we want to assess the quality of the markers: it is desirable to see at least two peaks, which would indicate the existence of cell type specific markers in our data.</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol type="1">
<li>Can we see any cell type specific markers, which are they?</li>
<li>Are there any non-cell type specific markers?</li>
<li>What can you notice about the distribution of markers across images?</li>
</ol>
</div>
<p>We assess whether the marker intensities require transformation or normalisation. This step is important for two main reasons:</p>
<ul>
<li><p><strong>Skewed distributions:</strong> Marker intensities are often right-skewed, which can distort downstream analyses such as clustering or dimensionality reduction.</p></li>
<li><p><strong>Inconsistent scales across images:</strong> Due to technical variation, the same marker may show very different intensity ranges across images. This can shift what’s considered “positive” or “negative” expression, making it difficult to label cells accurately.</p></li>
</ul>
<p>By applying transformation and normalisation, we aim to stabilise variance and bring the data onto a more comparable scale across images.</p>
<p><br></p>
<div class="panel-tabset">
<ul id="tabset-5" class="panel-tabset-tabby">
<li><a data-tabby-default="" href="#tabset-5-1">CD20</a></li>
<li><a href="#tabset-5-2">CD3</a></li>
<li><a href="#tabset-5-3">CD68</a></li>
<li><a href="#tabset-5-4">vWF_CD31</a></li>
<li><a href="#tabset-5-5">Vimentin</a></li>
<li><a href="#tabset-5-6">HER2</a></li>
</ul>
<ul class="nav nav-tabs" id="NA" role="tablist"><li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
</button></li></ul>
<div class="tab-content"><div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab">

<div id="tabset-5-2">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-12-2.png" width="672"></p>
</div>
<div id="tabset-5-3">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-12-3.png" width="672"></p>
</div>
<div id="tabset-5-4">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-12-4.png" width="672"></p>
</div>
<div id="tabset-5-5">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-12-5.png" width="672"></p>
</div>
<div id="tabset-5-6">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-12-6.png" width="672"></p>
</div>
</div></div>
</div>
<section class="level3"><h4 id="normalisation-of-marker-expression">Normalisation of marker expression<a class="anchor" aria-label="anchor" href="#normalisation-of-marker-expression"></a>
</h4>
<div>
<blockquote>
<p><strong>Tip</strong></p>
<p><strong>What we’re looking for</strong></p>
<ol type="1">
<li>Do the CD31+ and CD31- peaks clearly separate out in the density plot? To ensure that downstream analysis goes smoothly, we want our cell type specific markers to show 2 distinct peaks representing our CD31+ and CD31- cells. If we see 3 or more peaks where we don’t expect, this might be an indicator that further normalisation is required.</li>
<li>Are our CD31+ and CD31- peaks consistent across our images? We want to make sure that our density plots for CD3 are largely the same across images so that a CD3+ cell in one image is equivalent to a CD3+ cell in another image.</li>
</ol>
</blockquote>
</div>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># leave out the nuclei markers from our normalisation process</span></span>
<span><span class="va">useMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DNA1"</span>, <span class="st">"DNA2"</span>, <span class="st">"HH3"</span>, <span class="st">"HH3_total"</span>, <span class="st">"HH3_ph"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># transform and normalise the marker expression of each cell type</span></span>
<span><span class="va">data_sce</span> <span class="op">&lt;-</span> <span class="fu">normalizeCells</span><span class="op">(</span><span class="va">data_sce</span>,</span>
<span>                        markers <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>                        transformation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                        method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trim99"</span>, <span class="st">"minMax"</span>, <span class="st">"PC1"</span><span class="op">)</span>,</span>
<span>                        assayIn <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>                        imageID <span class="op">=</span> <span class="st">"metabricId"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">selected_patients</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">metabricId</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">cell_counts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span><span class="op">$</span><span class="va">metabricId</span><span class="op">)</span></span>
<span></span>
<span><span class="va">norm</span> <span class="op">=</span> <span class="fu">plot_marker_densities</span><span class="op">(</span><span class="va">data_sce</span><span class="op">[</span>,<span class="va">selected_patients</span><span class="op">]</span>, sample_id_col <span class="op">=</span> <span class="st">"metabricId"</span>, markers_to_plot <span class="op">=</span> <span class="st">"vWF_CD31"</span>, assay_name <span class="op">=</span> <span class="st">"norm"</span><span class="op">)</span><span class="op">$</span><span class="va">vWF_CD31</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Normalized Counts - CD31"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">log</span> <span class="op">=</span> <span class="fu">plot_marker_densities</span><span class="op">(</span><span class="va">data_sce</span><span class="op">[</span>,<span class="va">selected_patients</span><span class="op">]</span>, sample_id_col <span class="op">=</span> <span class="st">"metabricId"</span>, markers_to_plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"vWF_CD31"</span><span class="op">)</span>, assay_name <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span><span class="op">$</span><span class="va">vWF_CD31</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Log Counts - CD31"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggarrange</span><span class="op">(</span><span class="va">log</span>, <span class="va">norm</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In the plot above, the normalised data appeas more bimodal. We can observe one clear CD31- peak at around 0.50, and a CD31+ peak at 1.00. Image-level batch effects also appear to have been mitigated, since most peaks occur at around the same CD31 intensity.</p>
<div class="question">
<p><em>Discussion</em></p>
<ul>
<li>Do we have a sufficient amount of cell type specific markers?</li>
<li>Is the panel of genes sufficient for our study?</li>
</ul>
</div>
<p><br></p>
</section></section><section class="level2"><h3 id="qc-of-individual-samples">2.2: QC of individual samples<a class="anchor" aria-label="anchor" href="#qc-of-individual-samples"></a>
</h3>
<p>Here we plot the correlation of different markers for a subset of samples. This helps us determine whether a sample has low/high contamination of markers.</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol type="1">
<li>Which pairs of markers do we expect to be correlated/not correlated.</li>
<li>Is there any evidence of marker contamination for a given sample?</li>
</ol>
</div>
<p><br></p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coexp_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"data/coexp_df.rds"</span><span class="op">)</span></span>
<span><span class="va">marker_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD20"</span>, <span class="st">"CD3"</span>, <span class="st">"CD68"</span>, <span class="st">"vWF_CD31"</span>, <span class="st">"Vimentin"</span>, <span class="st">"HER2"</span><span class="op">)</span></span>
<span> </span>
<span><span class="co"># Proportion heatmaps</span></span>
<span><span class="va">heatmaps_prop</span> <span class="op">&lt;-</span> <span class="fu">plot_pairwise_heatmaps_per_sample</span><span class="op">(</span><span class="va">coexp_df</span>, <span class="va">marker_list</span>, stat <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="panel-tabset">
<ul id="tabset-6" class="panel-tabset-tabby">
<li><a data-tabby-default="" href="#tabset-6-1">MB-0002</a></li>
<li><a href="#tabset-6-2">MB-0064</a></li>
<li><a href="#tabset-6-3">MB-0128</a></li>
<li><a href="#tabset-6-4">MB-0130</a></li>
<li><a href="#tabset-6-5">MB-0132</a></li>
<li><a href="#tabset-6-6">MB-0136</a></li>
<li><a href="#tabset-6-7">MB-0138</a></li>
<li><a href="#tabset-6-8">MB-0142</a></li>
<li><a href="#tabset-6-9">MB-0145</a></li>
<li><a href="#tabset-6-10">MB-0154</a></li>
<li><a href="#tabset-6-11">MB-0168</a></li>
<li><a href="#tabset-6-12">MB-0175</a></li>
<li><a href="#tabset-6-13">MB-0180</a></li>
<li><a href="#tabset-6-14">MB-0192</a></li>
<li><a href="#tabset-6-15">MB-0201</a></li>
<li><a href="#tabset-6-16">MB-0208</a></li>
<li><a href="#tabset-6-17">MB-0225</a></li>
<li><a href="#tabset-6-18">MB-0227</a></li>
<li><a href="#tabset-6-19">MB-0232</a></li>
<li><a href="#tabset-6-20">MB-0240</a></li>
<li><a href="#tabset-6-21">MB-0245</a></li>
<li><a href="#tabset-6-22">MB-0246</a></li>
<li><a href="#tabset-6-23">MB-0249</a></li>
<li><a href="#tabset-6-24">MB-0252</a></li>
<li><a href="#tabset-6-25">MB-0255</a></li>
<li><a href="#tabset-6-26">MB-0256</a></li>
<li><a href="#tabset-6-27">MB-0258</a></li>
<li><a href="#tabset-6-28">MB-0260</a></li>
<li><a href="#tabset-6-29">MB-0263</a></li>
<li><a href="#tabset-6-30">MB-0275</a></li>
<li><a href="#tabset-6-31">MB-0308</a></li>
<li><a href="#tabset-6-32">MB-0318</a></li>
<li><a href="#tabset-6-33">MB-0320</a></li>
<li><a href="#tabset-6-34">MB-0347</a></li>
<li><a href="#tabset-6-35">MB-0351</a></li>
<li><a href="#tabset-6-36">MB-0367</a></li>
<li><a href="#tabset-6-37">MB-0377</a></li>
<li><a href="#tabset-6-38">MB-0383</a></li>
<li><a href="#tabset-6-39">MB-0394</a></li>
<li><a href="#tabset-6-40">MB-0397</a></li>
<li><a href="#tabset-6-41">MB-0400</a></li>
<li><a href="#tabset-6-42">MB-0401</a></li>
<li><a href="#tabset-6-43">MB-0405</a></li>
<li><a href="#tabset-6-44">MB-0410</a></li>
<li><a href="#tabset-6-45">MB-0420</a></li>
<li><a href="#tabset-6-46">MB-0439</a></li>
<li><a href="#tabset-6-47">MB-0451</a></li>
<li><a href="#tabset-6-48">MB-0454</a></li>
<li><a href="#tabset-6-49">MB-0455</a></li>
<li><a href="#tabset-6-50">MB-0461</a></li>
<li><a href="#tabset-6-51">MB-0463</a></li>
<li><a href="#tabset-6-52">MB-0467</a></li>
<li><a href="#tabset-6-53">MB-0470</a></li>
<li><a href="#tabset-6-54">MB-0475</a></li>
<li><a href="#tabset-6-55">MB-0480</a></li>
<li><a href="#tabset-6-56">MB-0481</a></li>
<li><a href="#tabset-6-57">MB-0487</a></li>
<li><a href="#tabset-6-58">MB-0498</a></li>
<li><a href="#tabset-6-59">MB-0508</a></li>
<li><a href="#tabset-6-60">MB-0516</a></li>
<li><a href="#tabset-6-61">MB-0518</a></li>
<li><a href="#tabset-6-62">MB-0520</a></li>
<li><a href="#tabset-6-63">MB-0531</a></li>
<li><a href="#tabset-6-64">MB-0536</a></li>
<li><a href="#tabset-6-65">MB-0571</a></li>
<li><a href="#tabset-6-66">MB-0584</a></li>
<li><a href="#tabset-6-67">MB-0591</a></li>
<li><a href="#tabset-6-68">MB-0596</a></li>
<li><a href="#tabset-6-69">MB-0604</a></li>
<li><a href="#tabset-6-70">MB-0605</a></li>
<li><a href="#tabset-6-71">MB-0628</a></li>
<li><a href="#tabset-6-72">MB-0635</a></li>
<li><a href="#tabset-6-73">MB-0636</a></li>
<li><a href="#tabset-6-74">MB-0642</a></li>
<li><a href="#tabset-6-75">MB-0661</a></li>
<li><a href="#tabset-6-76">MB-0662</a></li>
<li><a href="#tabset-6-77">MB-0663</a></li>
</ul>
<ul class="nav nav-tabs" id="NA" role="tablist"><li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-1.png" width="672"></p>
</button></li></ul>
<div class="tab-content"><div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab">

<div id="tabset-6-2">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-2.png" width="672"></p>
</div>
<div id="tabset-6-3">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-3.png" width="672"></p>
</div>
<div id="tabset-6-4">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-4.png" width="672"></p>
</div>
<div id="tabset-6-5">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-5.png" width="672"></p>
</div>
<div id="tabset-6-6">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-6.png" width="672"></p>
</div>
<div id="tabset-6-7">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-7.png" width="672"></p>
</div>
<div id="tabset-6-8">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-8.png" width="672"></p>
</div>
<div id="tabset-6-9">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-9.png" width="672"></p>
</div>
<div id="tabset-6-10">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-10.png" width="672"></p>
</div>
<div id="tabset-6-11">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-11.png" width="672"></p>
</div>
<div id="tabset-6-12">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-12.png" width="672"></p>
</div>
<div id="tabset-6-13">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-13.png" width="672"></p>
</div>
<div id="tabset-6-14">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-14.png" width="672"></p>
</div>
<div id="tabset-6-15">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-15.png" width="672"></p>
</div>
<div id="tabset-6-16">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-16.png" width="672"></p>
</div>
<div id="tabset-6-17">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-17.png" width="672"></p>
</div>
<div id="tabset-6-18">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-18.png" width="672"></p>
</div>
<div id="tabset-6-19">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-19.png" width="672"></p>
</div>
<div id="tabset-6-20">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-20.png" width="672"></p>
</div>
<div id="tabset-6-21">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-21.png" width="672"></p>
</div>
<div id="tabset-6-22">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-22.png" width="672"></p>
</div>
<div id="tabset-6-23">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-23.png" width="672"></p>
</div>
<div id="tabset-6-24">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-24.png" width="672"></p>
</div>
<div id="tabset-6-25">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-25.png" width="672"></p>
</div>
<div id="tabset-6-26">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-26.png" width="672"></p>
</div>
<div id="tabset-6-27">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-27.png" width="672"></p>
</div>
<div id="tabset-6-28">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-28.png" width="672"></p>
</div>
<div id="tabset-6-29">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-29.png" width="672"></p>
</div>
<div id="tabset-6-30">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-30.png" width="672"></p>
</div>
<div id="tabset-6-31">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-31.png" width="672"></p>
</div>
<div id="tabset-6-32">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-32.png" width="672"></p>
</div>
<div id="tabset-6-33">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-33.png" width="672"></p>
</div>
<div id="tabset-6-34">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-34.png" width="672"></p>
</div>
<div id="tabset-6-35">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-35.png" width="672"></p>
</div>
<div id="tabset-6-36">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-36.png" width="672"></p>
</div>
<div id="tabset-6-37">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-37.png" width="672"></p>
</div>
<div id="tabset-6-38">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-38.png" width="672"></p>
</div>
<div id="tabset-6-39">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-39.png" width="672"></p>
</div>
<div id="tabset-6-40">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-40.png" width="672"></p>
</div>
<div id="tabset-6-41">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-41.png" width="672"></p>
</div>
<div id="tabset-6-42">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-42.png" width="672"></p>
</div>
<div id="tabset-6-43">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-43.png" width="672"></p>
</div>
<div id="tabset-6-44">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-44.png" width="672"></p>
</div>
<div id="tabset-6-45">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-45.png" width="672"></p>
</div>
<div id="tabset-6-46">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-46.png" width="672"></p>
</div>
<div id="tabset-6-47">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-47.png" width="672"></p>
</div>
<div id="tabset-6-48">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-48.png" width="672"></p>
</div>
<div id="tabset-6-49">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-49.png" width="672"></p>
</div>
<div id="tabset-6-50">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-50.png" width="672"></p>
</div>
<div id="tabset-6-51">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-51.png" width="672"></p>
</div>
<div id="tabset-6-52">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-52.png" width="672"></p>
</div>
<div id="tabset-6-53">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-53.png" width="672"></p>
</div>
<div id="tabset-6-54">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-54.png" width="672"></p>
</div>
<div id="tabset-6-55">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-55.png" width="672"></p>
</div>
<div id="tabset-6-56">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-56.png" width="672"></p>
</div>
<div id="tabset-6-57">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-57.png" width="672"></p>
</div>
<div id="tabset-6-58">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-58.png" width="672"></p>
</div>
<div id="tabset-6-59">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-59.png" width="672"></p>
</div>
<div id="tabset-6-60">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-60.png" width="672"></p>
</div>
<div id="tabset-6-61">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-61.png" width="672"></p>
</div>
<div id="tabset-6-62">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-62.png" width="672"></p>
</div>
<div id="tabset-6-63">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-63.png" width="672"></p>
</div>
<div id="tabset-6-64">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-64.png" width="672"></p>
</div>
<div id="tabset-6-65">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-65.png" width="672"></p>
</div>
<div id="tabset-6-66">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-66.png" width="672"></p>
</div>
<div id="tabset-6-67">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-67.png" width="672"></p>
</div>
<div id="tabset-6-68">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-68.png" width="672"></p>
</div>
<div id="tabset-6-69">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-69.png" width="672"></p>
</div>
<div id="tabset-6-70">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-70.png" width="672"></p>
</div>
<div id="tabset-6-71">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-71.png" width="672"></p>
</div>
<div id="tabset-6-72">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-72.png" width="672"></p>
</div>
<div id="tabset-6-73">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-73.png" width="672"></p>
</div>
<div id="tabset-6-74">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-74.png" width="672"></p>
</div>
<div id="tabset-6-75">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-75.png" width="672"></p>
</div>
<div id="tabset-6-76">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-76.png" width="672"></p>
</div>
<div id="tabset-6-77">
<p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-15-77.png" width="672"></p>
</div>
</div></div>
</div>
</section><section class="level2"><h3 id="qc-of-study-as-a-whole">2.3: QC of study as a whole<a class="anchor" aria-label="anchor" href="#qc-of-study-as-a-whole"></a>
</h3>
<div class="question">
<p><strong>Questions</strong></p>
<ol type="1">
<li>Are there any samples you would like to remove from the data?</li>
<li>Which samples would you like to keep?</li>
</ol>
</div>
<p><br></p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_counts</span> <span class="op">&lt;-</span> <span class="va">IMC</span><span class="op">@</span><span class="va">colData</span> <span class="op">|&gt;</span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">metabricId</span>, name <span class="op">=</span> <span class="st">"cell_count"</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="co"># Count rows per metabricId</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">cell_count</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span></span>
<span><span class="va">mean_coexp_per_sample</span> <span class="op">&lt;-</span> <span class="va">coexp_df</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">cell_id</span>, <span class="op">-</span><span class="va">cell_type</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="co"># remove columns not related to prob values</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">metabricId</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_coexp_prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu">c_across</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">merged_df</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">mean_coexp_per_sample</span>, <span class="va">cell_counts</span>, by <span class="op">=</span> <span class="st">"metabricId"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Scatter plot</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">merged_df</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">cell_count</span>, y <span class="op">=</span> <span class="va">mean_coexp_prob</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span>, color <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">labs</span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Cell Count per Sample"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Mean Co-expression Probability"</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Mean Co-expression vs. Cell Count"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-16-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h3 id="does-my-data-agree-with-literature">2.4: Does my data agree with literature?<a class="anchor" aria-label="anchor" href="#does-my-data-agree-with-literature"></a>
</h3>
<p>Another way to QC our data is to explore whether the DE genes or associations we find matches current literature. For example, we have a breast cancer cohort. Breast cancer is very well studied and so it should be very easy to perform a DE analysis after pseudobulking our data to see if it makes findings in previous bulk RNA-seq studies of breast cancer cohorts. We can also do this at the single-cell level and compare with previous single-cell proteomic/RNA-seq studies of breast cancer cohorts. We won’t be running the code below. But here is an example of how one might perform a basic DE analysis using package <a href="https://academic.oup.com/nar/article/43/7/e47/2414268" class="external-link">limma</a>.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note: The same approach can be used for pseudobulk or single-cell level data. </span></span>
<span><span class="co"># Just change `logcounts(data_sce)`</span></span>
<span></span>
<span><span class="co"># We want low ki67 to be the reference level (baseline level) </span></span>
<span><span class="va">data_sce</span><span class="op">$</span><span class="va">ki67</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">$</span><span class="va">ki67</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"low_ki67"</span>, <span class="st">"high_ki67"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Here we specify the design matrix. We are specifying that Y are the expression values and X is ki67 status (but it could be any other variable - such as "good" or "bad" prognosis)</span></span>
<span><span class="co"># Y~X: Expression as a function of ki67 status. </span></span>
<span><span class="va">design_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">data_sce</span><span class="op">$</span><span class="va">ki67</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit model</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">lmFit</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">data_sce</span>, <span class="st">"norm"</span><span class="op">)</span>, design <span class="op">=</span> <span class="va">design_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Estimate variance and SE of coefficients</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">eBayes</span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tt</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="fl">2</span>, n <span class="op">=</span> <span class="fl">5</span>, adjust.method <span class="op">=</span> <span class="st">"BH"</span>, sort.by <span class="op">=</span> <span class="st">"p"</span><span class="op">)</span></span>
<span><span class="va">tt</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>                 logFC   AveExpr         t       P.Value     adj.P.Val
HH3_total   2.99807594 5.3783014 126.61566  0.000000e+00  0.000000e+00
HH3_ph      0.22158959 0.3549916  67.81304  0.000000e+00  0.000000e+00
Ki67        0.07903174 0.1129605  65.26755  0.000000e+00  0.000000e+00
E_cadherin  0.03507204 0.5389372  31.99330 4.085283e-223 3.881019e-222
H3K27me3   -0.03479189 0.4609833 -30.33584 6.094169e-201 4.631568e-200
                   B
HH3_total  7265.7780
HH3_ph     2222.8176
Ki67       2062.7347
E_cadherin  498.5455
H3K27me3    447.5419</code></pre>
</div>
</div>
</section><section class="level2"><h3 id="cell-type-classification">2.5: Cell Type Classification<a class="anchor" aria-label="anchor" href="#cell-type-classification"></a>
</h3>
<p>Here, we can also check whether the cell type annotations we have make sense. One way to do this is to check whether the marker expression levels agree with the cell type annotations. For example, we expect CD3 to be highly expressed in T cells, CD20 to be highly expressed in B cells, CD68 to be highly expressed in Macrophages, vWF_CD31 to be highly expressed in Endothelial cells, Vimentin to be highly expressed in Fibroblasts, and HER2 to be highly expressed in Tumor HER2+ cells.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate average expression per cell type</span></span>
<span><span class="va">avg_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">data_sce</span>, <span class="st">"norm"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                      by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>CellType <span class="op">=</span> <span class="va">IMC</span><span class="op">$</span><span class="va">high_level_category</span><span class="op">)</span>, </span>
<span>                      FUN <span class="op">=</span> <span class="va">mean</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"CellType"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">avg_expr</span> <span class="op">&lt;-</span> <span class="va">avg_expr</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD20"</span>, <span class="st">"CD3"</span>, <span class="st">"CD68"</span>, <span class="st">"CD45"</span>, </span>
<span>                         <span class="st">"vWF_CD31"</span>, <span class="st">"Vimentin"</span>, <span class="st">"HER2"</span><span class="op">)</span><span class="op">]</span>  <span class="co"># select only our mark</span></span>
<span></span>
<span><span class="co"># Plot heatmap of marker by celltype</span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">avg_expr</span>, scale <span class="op">=</span> <span class="st">"column"</span>, main <span class="op">=</span> <span class="st">"Average Marker Expression per Cell Type\n"</span>, </span>
<span>         color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">10000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-18-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="Discussion">
<p><em>Discussion</em></p>
<ul>
<li>How might be check whether these results make sense? What potential issues might we find?</li>
</ul>
</div>
<p><br></p>
</section></section><section class="section level2"><h2 id="part-3-exploring-spatial-data">Part 3: Exploring spatial data<a class="anchor" aria-label="anchor" href="#part-3-exploring-spatial-data"></a>
</h2>
<p>This section assumes pre-existing cell type annotations. For unannotated data, two primary annotation strategies are available:</p>
<ul>
<li>Unsupervised approach: cluster cells and identify marker genes for manual annotation.</li>
<li>Supervised approach: transfer labels from reference datasets using supervised learning/classification tools.</li>
</ul>
<p>For supervised annotation, we recommend <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC7306901/" class="external-link">scClassify</a>, a robust framework for cell-type classification. Below we visualise the data with and without spatial information before exploring the data further.</p>
<section class="level3"><h4 id="cell-types">Cell types<a class="anchor" aria-label="anchor" href="#cell-types"></a>
</h4>
<p>Following initial assessment of potential batch effects through sample-origin visualization, we now examine cell-type-specific clustering patterns. Distinct, biologically meaningful clusters should emerge for each annotated cell type. The presence of heterogeneous clusters containing unrelated cell types may indicate incomplete or inaccurate cell-type annotations.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">data_sce</span>, colour_by <span class="op">=</span> <span class="st">"description"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-19-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level3"><h4 id="spatial-structure">Spatial structure<a class="anchor" aria-label="anchor" href="#spatial-structure"></a>
</h4>
<p>The advantage with spatial omics is that we can examine the organisation of the cell-types as it occurs on the tissue slide. Here, we visualise one of the slides from a patient. We select a particular patient “MB-0002” and visualise the tissue sample from this patient using ggplot. Do we see any spatial patterning or does it look randomly distributed?</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># obtaining the meta data for this patient </span></span>
<span><span class="va">one_sample</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span>, <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span>  <span class="op">==</span> <span class="st">"MB-0002"</span><span class="op">]</span></span>
<span><span class="va">one_sample</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">one_sample</span><span class="op">)</span><span class="op">)</span></span>
<span> </span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">one_sample</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span>, y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">description</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_colour_tableau</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line<span class="op">=</span><span class="fu">element_line</span><span class="op">(</span>color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Original slide"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-20-1.png" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="question">
<p><strong>Questions</strong></p>
<ol type="1">
<li>What kinds of spatial information might be of interest given the question we’d like to answer?</li>
<li>How might we try to capture these spatial relationships?</li>
</ol>
</div>
<p><br></p>
<div class="panel-tabset">
<ul id="tabset-7" class="panel-tabset-tabby">
<li><a data-tabby-default="" href="#tabset-7-1">Spatial regions</a></li>
<li><a href="#tabset-7-2">Across individuals (Optional)</a></li>
<li><a href="#tabset-7-3">Specific regions (Optional)</a></li>
</ul>
<ul class="nav nav-tabs" id="NA" role="tablist"><li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<p>Background: One of the most common questions or analyses in spatial data is spatial domain detection. “Spatial domains” are regions within a tissue where cells share similar gene expression profiles and are physically clustered together. Here, we will use the terminology “spatial domain” and “regions” interchangeably. Most common analytical strategies involve spatial clustering, where different methods use different levels of information, such as gene expression data only, cell type information, and spatial coordinates.</p>
<p>The strategy we demonstrate here uses the <code>lisaClust</code> function, which uses cell type information to cluster cells into five distinct spatial regions. As a case study, we will compare individuals with good or poor prognosis and examine them graphically to see if any regions appear to be different between good or poor prognosis. We define: - Good prognosis as individuals with &gt; 10 years recurrence-free survival and - Poor prognosis as individuals with &lt; 5 years recurrence-free survival.</p>
<p>Below we visualise the spatial domain (regions) detection result based on one individual. Here we will use the terminology “spatial domain” and “regions” interchangeably.</p>
<p>Depending on the number of regions, it may be more useful to visualise the spatial regions either collectively in a single graph or separately in multiple graphs. To visualise it in a single graph, the <code>hatchingPlot()</code> function is used to create hatching patterns for representating spatial regions and cell-types.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract time to recurrence-free survival</span></span>
<span><span class="va">clinical</span><span class="op">$</span><span class="va">survivalgroup</span> <span class="op">&lt;-</span> <span class="st">"neither"</span></span>
<span></span>
<span><span class="co">## Define poor and good prognosis</span></span>
<span><span class="va">clinical</span><span class="op">$</span><span class="va">survivalgroup</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="va">clinical</span><span class="op">$</span><span class="va">timeRFS</span>  <span class="op">&lt;</span> <span class="fl">5</span><span class="op">*</span> <span class="fl">365</span><span class="op">)</span> <span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"poor"</span> </span>
<span><span class="va">clinical</span><span class="op">$</span><span class="va">survivalgroup</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="va">clinical</span><span class="op">$</span><span class="va">timeRFS</span>  <span class="op">&gt;</span> <span class="fl">10</span><span class="op">*</span> <span class="fl">365</span><span class="op">)</span> <span class="op">]</span> <span class="op">&lt;-</span>  <span class="st">"good"</span></span>
<span></span>
<span><span class="co">## To visualise it in a single graph</span></span>
<span><span class="fu">hatchingPlot</span><span class="op">(</span></span>
<span>  <span class="va">data_sce</span>,</span>
<span>  region <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>  imageID <span class="op">=</span> <span class="st">"metabricId"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"description"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Location_Center_X"</span>, <span class="st">"Location_Center_Y"</span><span class="op">)</span> <span class="op">)</span> </span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-21-1.png" width="576"></p>
</figure>
</div>
</div>
</div>
<p><br></p>
<p>We have written a small function <code>draw_region_clustering_result</code> to visualise the data separately in multiple graphs for the individual <code>MB-0002</code>.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_region_clustering_result</span><span class="op">(</span><span class="va">data_sce</span> , </span>
<span>                              sample <span class="op">=</span> <span class="st">"metabricId"</span> , </span>
<span>                              selected_sample <span class="op">=</span>  <span class="st">"MB-0002"</span> <span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>[[1]]</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-22-1.png" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
[[2]]</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-22-2.png" width="960"></p>
</figure>
</div>
</div>
</div>
<div class="question">
<p><em>Discussion</em></p>
<ul>
<li>If we’d like to examine multiple individuals, is there a better visualise this information?</li>
</ul>
</div>
<p><br></p>
</button></li></ul>
<div class="tab-content"><div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab">

<div id="tabset-7-2">
<p>We can better interpret the region output by summarising the proportion of each cell-type in a region across the individuals. We look at the composition of cell-types in each region and compare between prognostic outcomes.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_dotplot</span><span class="op">(</span><span class="va">data_sce</span>,  </span>
<span>             sample <span class="op">=</span> <span class="st">"metabricId"</span> , </span>
<span>             celltype <span class="op">=</span> <span class="st">"description"</span> ,  </span>
<span>             group<span class="op">=</span> <span class="st">"survivalgroup"</span> , </span>
<span>             group_of_interest <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"poor"</span> , <span class="st">"good"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'Var1', 'Var2'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-23-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-7-3">
<p>The number of sub-cell types increase considerably when we want to add spatial domain (region) information. To enhance clarity and facilitate understanding, it may be helpful to choose a predetermined region. The code generates a set of boxplots that enable the comparison of cell-type proportions between individuals with good and poor prognosis in <code>region_5</code>.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_selected_region_boxplot</span><span class="op">(</span><span class="va">data_sce</span>, </span>
<span>                             sample <span class="op">=</span> <span class="st">"metabricId"</span> , </span>
<span>                             celltype <span class="op">=</span><span class="st">"description"</span> , </span>
<span>                             group  <span class="op">=</span> <span class="st">"survivalgroup"</span>, </span>
<span>                             group_of_interest <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"poor"</span> , <span class="st">"good"</span><span class="op">)</span>, </span>
<span>                             select_region <span class="op">=</span> <span class="st">"region_5"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-24-1.png" width="1152"></p>
</figure>
</div>
</div>
</div>
</div>
</div></div>
</div>
</section></section><section class="section level2"><h2 id="part-4-feature-engineering-with-scfeatures">Part 4: Feature engineering with scFeatures<a class="anchor" aria-label="anchor" href="#part-4-feature-engineering-with-scfeatures"></a>
</h2>
<p>Here, we use scFeatures to generate molecular features for each individual using the <code>features x cells</code> matrices. These features are interpretable and can be used for downstream analyses. In general, <code>scFeatures</code> generates features across six categories:</p>
<ol type="i">
<li>Cell-type proportions.</li>
<li>Cell-type specific gene expressions.</li>
<li>Cell-type specific pathway expressions.</li>
<li>Cell-type interaction scores.</li>
<li>Aggregated gene expressions.</li>
<li>Spatial metrics: Nearest neighbour’s correlation, L statistics, and Moran’s I.</li>
</ol>
<p>The different types of features constructed enable a more comprehensive multi-view understanding of each individual. By default, the function will generate a total of 13 different types of features and are stored as 13 <code>samples x features</code> matrices, one for each feature type.</p>
<p>In this section, we will examine spatial information from two perspectives. Utilising spatial domain detection described in part 3, we select a specific region of interest and create molecular representations of that region for each individual (Section 4.1). Second, we will utilise spatial statistics to capture spatial relationships within the region of interest, such as Moran’s I (Section 4.2)</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">region</span></span>
<span></span>
<span><span class="co"># Define a series of sub-cell-types that is regional specific</span></span>
<span><span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">description</span> , <span class="st">"-"</span> , <span class="va">region</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<section class="level2"><h3 id="how-to-create-molecular-representations-of-individuals-for-an-roi">4.1 How to create molecular representations of individuals for an ROI?<a class="anchor" aria-label="anchor" href="#how-to-create-molecular-representations-of-individuals-for-an-roi"></a>
</h3>
<p>Here, we can consider regional information and the following code allows us to create cell-type specific features for each region. We use the function “paste0” to construct <strong>region-specific sub cell-types</strong> and name it as <code>celltype</code> in the R object <code>data_sce</code>. For simplicity, in this workshop, the variable <code>celltype</code> in the R object <code>data_sce</code> refers to region-specific sub-cell-types.</p>
<p>There are a total of 13 different types of features (feature_types) that you can choose to generate. The argument type refers the type of input data we have. This is either <code>scrna</code> (single-cell RNA-sequencing data), <code>spatial_p</code> (spatial proteomics data), or <code>spatial_t</code> (single-cell spatial data). In this section, we will ignore spatial information and generate non-spatial features, such as pseudobulking at the sample /cell type levels, overall expression, cell type proportions etc…</p>
<div class="panel-tabset">
<ul id="tabset-8" class="panel-tabset-tabby">
<li><a data-tabby-default="" href="#tabset-8-1">Generate features</a></li>
<li><a href="#tabset-8-2">ROI</a></li>
<li><a href="#tabset-8-3">scFeatures code (Optional)</a></li>
<li><a href="#tabset-8-4">scFeatures output (Optional)</a></li>
</ul>
<ul class="nav nav-tabs" id="NA" role="tablist"><li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<p>Suppose that we are interested in determining the proportion of cell-types within each region for each individual. It is necessary to specify <code>type = spatial_p</code> to reflect that we have spatial proteomics data and <code>feature_types = proportion_raw</code> to indicate we intend to calculate cell-type proportions for each of the region-specific cell-types. Suppose we are only interested in the molecular representation of <code>HR- CK7+</code> within individuals within region 5.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [A] The next few lines extract specific information from data_sce as input to scFeatures. </span></span>
<span><span class="co">## Extract the expression matrix from data_sce</span></span>
<span><span class="va">IMCmatrix</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">data_sce</span>, <span class="st">"norm"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract the sample information </span></span>
<span><span class="co">## append the condition to the individuals so we can easily retrieve the individuals condition </span></span>
<span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> </span>
<span><span class="va">cond</span>  <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">metabricId</span><span class="op">)</span>, <span class="op">]</span><span class="op">$</span><span class="va">survivalgroup</span></span>
<span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"_cond_"</span>, <span class="va">cond</span> <span class="op">)</span> </span>
<span></span>
<span><span class="co">## Extract the region-specific sub-cell-types</span></span>
<span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span></span>
<span></span>
<span><span class="co">## Extract the spatial coordinates</span></span>
<span><span class="va">spatialCoords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_X</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### [B] Running scFeatures</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu">scFeatures</span><span class="op">(</span><span class="va">IMCmatrix</span>, </span>
<span>                                sample <span class="op">=</span> <span class="va">sample</span>, </span>
<span>                                celltype <span class="op">=</span> <span class="va">celltype</span>, </span>
<span>                                spatialCoords <span class="op">=</span> <span class="va">spatialCoords</span>,</span>
<span>                                feature_types <span class="op">=</span> <span class="st">"proportion_raw"</span>, </span>
<span>                                type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span></code></pre></div>
</details>
</div>
</button></li></ul>
<div class="tab-content"><div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab">

<div id="tabset-8-2">
<div class="question">
<p><strong>Question</strong></p>
<ol type="1">
<li>Are there any regions that are associated with “good” and “poor” prognosis?</li>
<li>Is this the right way to visualise the results?</li>
</ol>
</div>
<p><br></p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [A] The next few lines extract specific information from data_sce as input to scFeatures. </span></span>
<span><span class="co">## Select the HR- CK7+-region sub-cell-type </span></span>
<span></span>
<span><span class="co"># There are different ways you can use `scFeatures` to generate molecular representations for individuals and it requires the following information for spatial data.</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># data,\</span></span>
<span><span class="co"># sample,</span></span>
<span><span class="co"># X coordinates,</span></span>
<span><span class="co"># Y coordinates,</span></span>
<span><span class="co"># feature_types, and</span></span>
<span><span class="co"># type</span></span>
<span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"HR- CK7+-region"</span> , <span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span>, fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">selected_data</span> <span class="op">&lt;-</span> <span class="va">IMCmatrix</span><span class="op">[</span>, <span class="va">index</span><span class="op">]</span></span>
<span><span class="va">selected_sample</span> <span class="op">&lt;-</span> <span class="va">sample</span><span class="op">[</span><span class="va">index</span><span class="op">]</span></span>
<span><span class="va">selected_celltype</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span> <span class="va">index</span><span class="op">]</span> </span>
<span><span class="va">selected_spatialCoords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_X</span><span class="op">[</span><span class="va">index</span><span class="op">]</span>, </span>
<span>                               <span class="fu">colData</span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">[</span><span class="va">index</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### [B] Running scFeatures</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu">scFeatures</span><span class="op">(</span> <span class="va">selected_data</span>, </span>
<span>                                 sample <span class="op">=</span> <span class="va">selected_sample</span>, </span>
<span>                                 celltype <span class="op">=</span> <span class="va">selected_celltype</span>,</span>
<span>                                 spatialCoords <span class="op">=</span> <span class="va">selected_spatialCoords</span>,</span>
<span>                                 feature_types <span class="op">=</span> <span class="st">"proportion_raw"</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span>
<span></span>
<span><span class="co">### [C] Visualize the regional composition makeup for each individual for HR- CK7+ and HR- CK7-</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">proportion_raw</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span></span>
<span><span class="fu">plot_barplot</span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Proportion raw feature"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"proportion\n"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-27-1.png" width="1920"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-8-3">
<p>The code below enable us to generate all feature types for all cell-types in a line. Due to limitations with today’s computational capacity, <strong>Please DO NOT run it in today’s workshop, it will crash your system</strong>.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># here, we specify that this is a spatial proteomics data</span></span>
<span><span class="co"># scFeatures support parallel computation to speed up the process </span></span>
<span><span class="co"># scfeatures_result &lt;- scFeatures(IMCmatrix, </span></span>
<span><span class="co">#                                 type = "spatial_p",</span></span>
<span><span class="co">#                                 sample = sample, </span></span>
<span><span class="co">#                                 celltype = celltype, </span></span>
<span><span class="co">#                                 spatialCoords = spatialCoords,</span></span>
<span><span class="co">#                                 ncores = 32)</span></span></code></pre></div>
</details>
</div>
</div>
<div id="tabset-8-4">
<p>Assuming you have already generated a collection of molecular representation for individuals, please load the prepared RDS file <code>scfeatures_result.RDS</code>. Again, you can remind yourself that all generated feature types are stored in a matrix of <code>samples x features</code>.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Upload pre-generated RDS file</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"data/scfeatures_result.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># What are the features and the dimensions of features matrices that we have generated?</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">scfeatures_result</span>, <span class="va">dim</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>$proportion_raw
[1]  77 110

$proportion_logit
[1]  77 110

$proportion_ratio
[1]   77 5995

$gene_mean_celltype
[1]   77 4180

$gene_prop_celltype
[1]   77 4180

$gene_cor_celltype
[1]    77 27958

$gene_mean_bulk
[1] 77 38

$gene_prop_bulk
[1] 77 38

$gene_cor_bulk
[1]  77 703

$L_stats
[1]   77 3842

$celltype_interaction
[1]   77 4393

$morans_I
[1] 77 38

$nn_correlation
[1] 77 38</code></pre>
</div>
</div>
</div>
</div></div>
</div>
</section><section class="level2"><h3 id="how-can-we-represent-spatial-information-and-relationships-for-a-given-individual">4.2 How can we represent spatial information and relationships for a given individual?<a class="anchor" aria-label="anchor" href="#how-can-we-represent-spatial-information-and-relationships-for-a-given-individual"></a>
</h3>
<p>We will now look at the spatial statistic output by <code>scFeatures</code> and qualitively assess whether there is any association between between these statistics and the “good” and “bad” prognosis groups.</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol type="1">
<li>What kind of information would spatial statistics provide?</li>
<li>Are there any differences in the distribution of spatial statistics between the “good” and “bad” prognosis groups?</li>
</ol>
</div>
<p><br></p>
<div class="panel-tabset">
<ul id="tabset-9" class="panel-tabset-tabby">
<li><a data-tabby-default="" href="#tabset-9-1">Nearest neighbour correlation</a></li>
<li><a href="#tabset-9-2">L statistics</a></li>
<li><a href="#tabset-9-3">Moran’s I</a></li>
</ul>
<ul class="nav nav-tabs" id="NA" role="tablist"><li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<p>This metric measures the correlation of proteins/genes between a cell and its nearest neighbour cell.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">nn_correlation</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line<span class="op">=</span><span class="fu">element_line</span><span class="op">(</span>color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Nearest neighbour correlation feature"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-30-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</button></li></ul>
<div class="tab-content"><div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab">

<div id="tabset-9-2">
<p>The L function is a spatial statistic used to assess the spatial distribution of cell-types. It assesses the significance of cell-cell interactions, by calculating the density of a cell-type with other cell-types within a certain radius. High values indicate spatial association (co-localisation), low values indicate spatial avoidance.For clarity, we plot the L statistics for a subset of regions.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">L_stats</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^T cells-region4_HR"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span><span class="op">==</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span>,<span class="va">idx</span><span class="op">]</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line<span class="op">=</span><span class="fu">element_line</span><span class="op">(</span>color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"L statistics"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-31-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-9-3">
<p>Moran’s I is a spatial autocorrelation statistic based on both location and values. It quantifies whether similar values tend to occur near each other or are dispersed.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">morans_I</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line<span class="op">=</span><span class="fu">element_line</span><span class="op">(</span>color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Moran's I"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-32-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div></div>
</div>
</section></section><section class="section level2"><h2 id="part-5-disease-classification-with-classifyr-presentation">Part 5: Disease classification with ClassifyR [Presentation]<a class="anchor" aria-label="anchor" href="#part-5-disease-classification-with-classifyr-presentation"></a>
</h2>
<p>Recurrence risk estimation is a fundamental concern in medical research, particularly in the context of patient survival analysis. In this section, we will estimate recurrence risk using the molecular representation of individuals generated from <code>scFeatures</code> to build a survival model. We will use classifyR to build the survival model. The patient outcome is time-to-event, so, by default, ClassifyR will use Cox proportional hazards ranking to choose a set of features and also Cox proportional hazards to predict risk scores. We will also demonstrate other available models in <code>ClassifyR</code>.</p>
<div class="question">
<p><strong>Questions</strong></p>
<ol type="1">
<li>Are spatial features globally informative in predicting survival?</li>
<li>If not, for which individuals is it important in predicting survival?</li>
</ol>
</div>
<p><br></p>
<div class="panel-tabset">
<ul id="tabset-10" class="panel-tabset-tabby">
<li><a data-tabby-default="" href="#tabset-10-1">Building a survival model (Optional)</a></li>
<li><a href="#tabset-10-2">Model performance (Optional)</a></li>
<li><a href="#tabset-10-3">Does spatial information matter?</a></li>
</ul>
<ul class="nav nav-tabs" id="NA" role="tablist"><li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#NA" id="NA-tab" type="button" role="tab" aria-controls="NA" aria-selected="true" class="active nav-link">
<p>Recall in the previous section that we have stored the 13 matrices of different feature types in a list. Instead of manually retrieving each matrix from the list to build separate models, classifyR can directly take a list of matrices as an input and run a repeated cross-validation model on each matrix individually. Below, we run 5 repeats of 5-fold cross-validation. A high score indicates prognosis of a worse outcome than a lower risk score. Although we have provided the code below, to save time, <strong>just load the prepared RDS file <code>classifyr_result_IMC.rds</code></strong> and we will focus on the interpretation in this workshop.</p>
<pre><code><span><span class="co"># We use the following variables:</span></span>
<span><span class="co"># timeRFS: "Time to Recurrence-Free Survival." It is the time period until recurrence occurs.</span></span>
<span><span class="co"># eventRFS: "Event in Recurrence-Free Survival."It indicates whether the event has occurred.</span></span>
<span><span class="co"># Breast.Tumour.Laterality: Laterality of tumors, eg, whether the tumor is located in left or right.</span></span>
<span><span class="co"># ER.Status: Whether the tumor is ER positive or ER negative.</span></span>
<span><span class="co"># Inferred.Menopausal.State: of the patient.</span></span>
<span><span class="co"># Grade: of the tumor.</span></span>
<span><span class="co"># Size: of the tumor.</span></span>
<span></span>
<span><span class="va">usefulFeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Breast.Tumour.Laterality"</span>, <span class="st">"ER.Status"</span>, <span class="st">"Inferred.Menopausal.State"</span>, <span class="st">"Grade"</span>, <span class="st">"Size"</span><span class="op">)</span></span>
<span><span class="va">nFeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">scfeatures_result</span>, <span class="kw">function</span><span class="op">(</span><span class="va">metaFeature</span><span class="op">)</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">clinicalAndOmics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="va">clinical</span><span class="op">)</span>, <span class="va">scfeatures_result</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### generate classfyr result</span></span>
<span><span class="va">classifyr_result_IMC</span> <span class="op">&lt;-</span> <span class="fu">crossValidate</span><span class="op">(</span><span class="va">clinicalAndOmics</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"timeRFS"</span>, <span class="st">"eventRFS"</span><span class="op">)</span>,</span>
<span>                    extraParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>prepare <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>useFeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="va">usefulFeatures</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                    nFeatures <span class="op">=</span> <span class="va">nFeatures</span>, nFolds <span class="op">=</span> <span class="fl">5</span>, nRepeats <span class="op">=</span> <span class="fl">5</span>, nCores <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">classifyr_result_IMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"data/classifyr_result_IMC.rds"</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<p>Cox proportional hazards is a classical statistical method, as opposed to machine learning methods like Random survival forest. These machine learning methods can build remarkably complex relationships between features, however their running time can be much longer than Cox proportional hazards. We use feature selection to limit the number of features considered to at most 100 per metafeature and to save time, <strong>you can just load the prepared RDS file.</strong> We will compare the predictive performance between these methods.</p>
<pre><code><span><span class="va">nFeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">metaFeature</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">metaFeature</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">survForestCV</span> <span class="op">&lt;-</span> <span class="fu">crossValidate</span><span class="op">(</span><span class="va">clinicalAndOmics</span>, <span class="va">outcome</span>, nFeatures <span class="op">=</span> <span class="va">nFeatures</span>,</span>
<span>                classifier <span class="op">=</span> <span class="st">"randomForest"</span>,</span>
<span>                nFolds <span class="op">=</span> <span class="fl">5</span>, nRepeats <span class="op">=</span> <span class="fl">5</span>, nCores <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">survForestCV</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"data/survForestCV.RDS"</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
</button></li></ul>
<div class="tab-content"><div class="active  tab-pane" id="NA" role="tabpanel" aria-labelledby="NA-tab">

<div id="tabset-10-2">
<p>To examine the distribution of prognostic performance, use <code>performancePlot</code>. Currently, the only metric for time-to-event data is C-index and that will automatically be used because the predictive model type is tracked inside of the result objects.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Make axis label 45 degree to improve readiability </span></span>
<span><span class="va">tilt</span> <span class="op">&lt;-</span> <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Putting two sets of cross-validation results together</span></span>
<span><span class="va">multiresults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">classifyr_result_IMC</span>, <span class="va">survForestCV</span><span class="op">)</span></span>
<span><span class="va">ordering</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clinical"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">performancePlot</span><span class="op">(</span><span class="va">multiresults</span>,</span>
<span>                characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span>, </span>
<span>                                           row <span class="op">=</span> <span class="st">"Classifier Name"</span><span class="op">)</span>,</span>
<span>                orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="va">ordering</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>                <span class="va">tilt</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-35-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Note how the resultant plot is a <code>ggplot2</code> object and can be further modified. The same code could be used for a categorical classifier because the random forest implementation provided by the <code>ranger</code> package has the same interface for both. We will examine feature selection stability with <code>selectionPlot</code>.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">selectionPlot</span><span class="op">(</span><span class="va">multiresults</span>,</span>
<span>                characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span>, row <span class="op">=</span> <span class="st">"Classifier Name"</span><span class="op">)</span>,</span>
<span>                orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="va">ordering</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="va">tilt</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-36-1.png" width="672"></p>
</figure>
</div>
</div>
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">distribution</span><span class="op">(</span><span class="va">classifyr_result_IMC</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>     assay                   feature proportion
1 clinical Inferred.Menopausal.State          1</code></pre>
</div>
</div>
</div>
<div id="tabset-10-3">
<p>Does each individual require a different collection of features? Using <code>samplesMetricMap</code> compare the per-sample C-index for Cox models for all kinds of metafeatures.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span></span>
<span><span class="fu">samplesMetricMap</span><span class="op">(</span><span class="va">classifyr_result_IMC</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-37-1.png" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>TableGrob (2 x 1) "arrange": 2 grobs
  z     cells    name                 grob
1 1 (2-2,1-1) arrange       gtable[layout]
2 2 (1-1,1-1) arrange text[GRID.text.9237]</code></pre>
</div>
</div>
</div>
</div></div>
</div>
</section><section class="section level2"><h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<section class="level2"><h3 id="explanation-of-spatial-features">Explanation of spatial features<a class="anchor" aria-label="anchor" href="#explanation-of-spatial-features"></a>
</h3>
<ul>
<li>L function:</li>
</ul>
<p>The L function is a spatial statistic used to assess the spatial distribution of cell-types. It assesses the significance of cell-cell interactions, by calculating the density of a cell-type with other cell-types within a certain radius. High values indicate spatial association (co-localisation), low values indicate spatial avoidance.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tableau_palette</span> <span class="op">&lt;-</span> <span class="fu">scale_colour_tableau</span><span class="op">(</span> palette <span class="op">=</span> <span class="st">"Tableau 20"</span><span class="op">)</span></span>
<span><span class="va">color_codes</span> <span class="op">&lt;-</span> <span class="va">tableau_palette</span><span class="op">$</span><span class="fu">palette</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co"># select one patient </span></span>
<span><span class="va">one_sample</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span> , <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0128"</span>  <span class="op">]</span></span>
<span><span class="va">one_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu">colData</span><span class="op">(</span><span class="va">one_sample</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">one_sample</span><span class="op">$</span><span class="va">description</span></span>
<span></span>
<span><span class="co"># select certain cell types to examine the interaction </span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span>  <span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span>  <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cells"</span>, <span class="st">"Fibroblasts"</span><span class="op">)</span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span><span class="op">!</span><span class="va">index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"others"</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span><span class="fu">ggplot</span><span class="op">(</span> <span class="va">one_sample</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">celltype</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_codes</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span> <span class="st">"Patient MB-0128 - high L value with \n B cells interacting Fibroblasts"</span><span class="op">)</span></span>
<span> </span>
<span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">one_sample</span><span class="op">$</span><span class="va">description</span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span>  <span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span>  <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Fibroblasts"</span>, <span class="st">"HR- CK7-"</span><span class="op">)</span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span><span class="op">!</span><span class="va">index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"others"</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span> <span class="va">one_sample</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">celltype</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_codes</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span> <span class="st">"Patient MB-0128 - low L value with \n B cells interacting HR_ CK7"</span><span class="op">)</span></span>
<span> </span>
<span><span class="fu">ggarrange</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-38-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Cell type interaction composition:</li>
</ul>
<p>We calculate the nearest neighbours of each cell and then calculate the pairs of cell-types based on the nearest neighbour. This allows us to summarise it into a cell-type interaction composition.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">one_sample</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span> , <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0263"</span>  <span class="op">]</span></span>
<span><span class="va">one_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu">colData</span><span class="op">(</span><span class="va">one_sample</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">one_sample</span><span class="op">$</span><span class="va">description</span></span>
<span> </span>
<span><span class="va">a</span> <span class="op">&lt;-</span><span class="fu">ggplot</span><span class="op">(</span> <span class="va">one_sample</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">celltype</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_codes</span><span class="op">)</span>  <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Patient MB-0263"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">feature</span>  <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">celltype_interaction</span></span>
<span><span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span> <span class="va">feature</span><span class="op">[</span> <span class="st">"MB-0263_cond_poor"</span> , <span class="op">]</span><span class="op">)</span>  <span class="op">)</span></span>
<span><span class="va">to_plot</span><span class="op">$</span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"value"</span>, <span class="st">"celltype interaction composition"</span><span class="op">)</span></span>
<span> </span>
<span><span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="va">to_plot</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">$</span><span class="va">value</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="va">to_plot</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span></span>
<span><span class="va">to_plot</span><span class="op">$</span><span class="va">`celltype interaction composition`</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">$</span><span class="va">`celltype interaction composition`</span>, levels <span class="op">=</span> <span class="va">to_plot</span><span class="op">$</span><span class="va">`celltype interaction composition`</span><span class="op">)</span></span>
<span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">to_plot</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span>  <span class="va">`celltype interaction composition`</span>  ,  y <span class="op">=</span> <span class="va">value</span>, fill<span class="op">=</span><span class="va">`celltype interaction composition`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_bar</span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span> <span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"Major cell-type interactions"</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu">ggarrange</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-39-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Moran’s I:</li>
</ul>
<p>Moran’s I is a spatial autocorrelation statistic based on both location and values. It quantifies whether similar values tend to occur near each other or are dispersed.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">high</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span><span class="st">"Ki67"</span>, <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0132"</span>  <span class="op">]</span></span>
<span><span class="va">high_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu">colData</span><span class="op">(</span><span class="va">high</span><span class="op">)</span> <span class="op">)</span> </span>
<span><span class="va">high_meta</span><span class="op">$</span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span> <span class="va">high</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">low</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span><span class="st">"Ki67"</span>,  <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0249"</span> <span class="op">]</span></span>
<span><span class="va">low_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu">colData</span><span class="op">(</span><span class="va">low</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">low_meta</span><span class="op">$</span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">low</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">high_meta</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_colour_viridis_c</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Patient MB-0132 - high Moran's I in Ki67"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">low_meta</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_colour_viridis_c</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Patient MB-0249 - low Moran's I in Ki67"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggarrange</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-40-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Nearest Neighbor Correlation:</li>
</ul>
<p>This metric measures the correlation of proteins/genes between a cell and its nearest neighbour cell.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_nncorrelation</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">thissample</span> , <span class="va">thisprotein</span><span class="op">)</span><span class="op">{</span></span>
<span>   </span>
<span>       <span class="va">sample_name</span> <span class="op">&lt;-</span> <span class="va">thissample</span></span>
<span>       <span class="va">thissample</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span>, <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span>     <span class="va">sample_name</span><span class="op">]</span></span>
<span>    </span>
<span>      </span>
<span>      <span class="va">exprsMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">)</span></span>
<span>     </span>
<span>    <span class="co"># calculate NN correlation </span></span>
<span>    <span class="va">cell_points_cts</span> <span class="op">&lt;-</span> <span class="fu">spatstat.geom</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spatstat.geom/man/ppp.html" class="external-link">ppp</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_X</span> <span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span>,</span>
<span>            check <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            xrange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_X</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_X</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="op">)</span>,</span>
<span>            yrange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="op">)</span>,</span>
<span>            marks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">exprsMat</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>    </span>
<span>     <span class="va">value</span> <span class="op">&lt;-</span>  <span class="fu">spatstat.explore</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spatstat.explore/man/nncorr.html" class="external-link">nncorr</a></span><span class="op">(</span><span class="va">cell_points_cts</span><span class="op">)</span><span class="op">[</span><span class="st">"correlation"</span>, <span class="op">]</span></span>
<span>      <span class="va">value</span> <span class="op">&lt;-</span>  <span class="va">value</span><span class="op">[</span>  <span class="va">thisprotein</span><span class="op">]</span></span>
<span>     </span>
<span>    <span class="co"># Find the indices of the two nearest neighbors for each cell</span></span>
<span>    <span class="va">nn_indices</span> <span class="op">&lt;-</span> <span class="fu">nnwhich</span><span class="op">(</span><span class="va">cell_points_cts</span>, k <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">protein</span> <span class="op">&lt;-</span>  <span class="va">thisprotein</span></span>
<span>    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>thiscell_exprs  <span class="op">=</span> <span class="va">exprsMat</span><span class="op">[</span><span class="va">protein</span>, <span class="op">]</span> , exprs <span class="op">=</span>  <span class="va">exprsMat</span><span class="op">[</span><span class="va">protein</span>,<span class="va">nn_indices</span> <span class="op">]</span><span class="op">)</span></span>
<span>    </span>
<span>   <span class="va">p</span> <span class="op">&lt;-</span>  <span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span>, <span class="fu">aes</span><span class="op">(</span> x <span class="op">=</span><span class="va">thiscell_exprs</span> ,  y <span class="op">=</span> <span class="va">exprs</span> , colour <span class="op">=</span>  <span class="va">exprs</span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span> <span class="st">"Patient "</span>, <span class="va">sample_name</span> ,  <span class="st">" nn_corr = "</span> ,  <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">value</span>, <span class="fl">2</span><span class="op">)</span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_colour_viridis_c</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">xlab</span><span class="op">(</span><span class="st">"This cell expression"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"Neighbouring cell expression"</span><span class="op">)</span></span>
<span>   </span>
<span>   <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span> <span class="op">(</span><span class="va">p</span> <span class="op">)</span> </span>
<span></span>
<span><span class="op">}</span></span>
<span></span>
<span>    </span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">plot_nncorrelation</span><span class="op">(</span><span class="st">"MB-0605"</span>,  <span class="st">"HER2"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">plot_nncorrelation</span><span class="op">(</span><span class="st">"MB-0258"</span>,  <span class="st">"HER2"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggarrange</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div>
<figure><p><img src="NUS_workshop_v1_files/figure-html/unnamed-chunk-41-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h3 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h3>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.1 (2025-06-13)
Platform: x86_64-pc-linux-gnu
Running under: Debian GNU/Linux 13 (trixie)

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.29.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: Australia/Sydney
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets
[8] methods   base

other attached packages:
 [1] cytomapper_1.20.0           EBImage_4.50.0
 [3] purrr_1.1.0                 simpleSeg_1.10.1
 [5] naniar_1.1.0                reshape_0.8.10
 [7] scran_1.36.0                scater_1.36.0
 [9] scuttle_1.18.0              spatstat_3.4-0
[11] spatstat.linnet_3.3-1       spatstat.model_3.4-0
[13] rpart_4.1.24                spatstat.explore_3.5-2
[15] nlme_3.1-168                spatstat.random_3.4-1
[17] spatstat.geom_3.5-0         spatstat.univar_3.1-4
[19] spatstat.data_3.1-8         survminer_0.5.1
[21] ggpubr_0.6.1                tidyr_1.3.1
[23] scattermore_1.2             plotly_4.11.0
[25] limma_3.64.3                dplyr_1.1.4
[27] spicyR_1.20.2               pheatmap_1.0.13
[29] ggthemes_5.1.0              lisaClust_1.16.0
[31] ClassifyR_3.12.5            survival_3.8-3
[33] BiocParallel_1.42.2         MultiAssayExperiment_1.34.0
[35] scFeatures_1.8.0            ggplot2_4.0.0
[37] SingleCellExperiment_1.30.1 SummarizedExperiment_1.38.1
[39] Biobase_2.68.0              GenomicRanges_1.60.0
[41] GenomeInfoDb_1.44.2         IRanges_2.42.0
[43] S4Vectors_0.46.0            BiocGenerics_0.54.1
[45] generics_0.1.4              MatrixGenerics_1.20.0
[47] matrixStats_1.5.0

loaded via a namespace (and not attached):
  [1] SpatialExperiment_1.18.1   R.methodsS3_1.8.2
  [3] dichromat_2.0-0.1          GSEABase_1.70.0
  [5] tiff_0.1-12                dcanr_1.24.0
  [7] EnsDb.Mmusculus.v79_2.99.0 goftest_1.2-3
  [9] Biostrings_2.76.0          HDF5Array_1.36.0
 [11] vctrs_0.6.5                digest_0.6.37
 [13] png_0.1-8                  shape_1.4.6.1
 [15] EnsDb.Hsapiens.v79_2.99.0  BiocBaseUtils_1.10.0
 [17] ggrepel_0.9.6              deldir_2.0-4
 [19] magick_2.9.0               MASS_7.3-65
 [21] reshape2_1.4.4             httpuv_1.6.16
 [23] foreach_1.5.2              withr_3.0.2
 [25] xfun_0.53                  doRNG_1.8.6.2
 [27] memoise_2.0.1              proxyC_0.5.2
 [29] ggbeeswarm_0.7.2           systemfonts_1.3.1
 [31] zoo_1.8-14                 GlobalOptions_0.1.2
 [33] gtools_3.9.5               V8_8.0.1
 [35] SingleCellSignalR_1.20.0   R.oo_1.27.1
 [37] Formula_1.2-5              promises_1.3.3
 [39] KEGGREST_1.48.1            httr_1.4.7
 [41] rstatix_0.7.2              restfulr_0.0.16
 [43] rhdf5filters_1.20.0        rhdf5_2.52.1
 [45] rstudioapi_0.17.1          UCSC.utils_1.4.0
 [47] concaveman_1.2.0           babelgene_22.9
 [49] curl_7.0.0                 ScaledMatrix_1.16.0
 [51] h5mread_1.0.1              polyclip_1.10-7
 [53] GenomeInfoDbData_1.2.14    SparseArray_1.8.1
 [55] fftwtools_0.9-11           xtable_1.8-4
 [57] stringr_1.5.2              evaluate_1.0.5
 [59] S4Arrays_1.8.1             irlba_2.3.5.1
 [61] colorspace_2.1-2           magrittr_2.0.4
 [63] later_1.4.4                viridis_0.6.5
 [65] lattice_0.22-7             cowplot_1.2.0
 [67] XML_3.99-0.19              ggupset_0.4.1
 [69] svgPanZoom_0.3.4           class_7.3-23
 [71] pillar_1.11.1              iterators_1.0.14
 [73] caTools_1.18.3             compiler_4.5.1
 [75] beachmat_2.24.0            stringi_1.8.7
 [77] tensor_1.5.1               minqa_1.2.8
 [79] GenomicAlignments_1.44.0   plyr_1.8.9
 [81] msigdbr_25.1.1             crayon_1.5.3
 [83] abind_1.4-8                BiocIO_1.18.0
 [85] sp_2.2-0                   locfit_1.5-9.12
 [87] terra_1.8-60               bit_4.6.0
 [89] textshaping_1.0.3          codetools_0.2-20
 [91] BiocSingular_1.24.0        mime_0.13
 [93] multtest_2.64.0            splines_4.5.1
 [95] circlize_0.4.16            Rcpp_1.1.0
 [97] sparseMatrixStats_1.20.0   knitr_1.50
 [99] blob_1.2.4                 AnnotationFilter_1.32.0
[101] lme4_1.1-37                nnls_1.6
[103] DelayedMatrixStats_1.30.0  Rdpack_2.6.4
[105] GSVA_2.2.0                 ggsignif_0.6.4
[107] tibble_3.3.0               Matrix_1.7-4
[109] scam_1.2-19                statmod_1.5.0
[111] svglite_2.2.1              visdat_0.6.0
[113] tweenr_2.0.3               pkgconfig_2.0.3
[115] tools_4.5.1                cachem_1.1.0
[117] rbibutils_2.3              RSQLite_2.4.3
[119] viridisLite_0.4.2          DBI_1.2.3
[121] numDeriv_2016.8-1.1        fastmap_1.2.0
[123] rmarkdown_2.30             scales_1.4.0
[125] shinydashboard_0.7.3       Rsamtools_2.24.1
[127] broom_1.0.10               BiocManager_1.30.26
[129] graph_1.86.0               carData_3.0-5
[131] farver_2.1.2               reformulas_0.4.1
[133] mgcv_1.9-3                 yaml_2.3.10
[135] rtracklayer_1.68.0         cli_3.6.5
[137] lifecycle_1.0.4            bluster_1.18.0
[139] backports_1.5.0            annotate_1.86.1
[141] gtable_0.3.6               rjson_0.2.23
[143] parallel_4.5.1             ape_5.8-1
[145] jsonlite_2.0.0             edgeR_4.6.3
[147] bitops_1.0-9               bit64_4.6.0-1
[149] assertthat_0.2.1           Rtsne_0.17
[151] spatstat.utils_3.1-5       BiocNeighbors_2.2.0
[153] bdsmatrix_1.3-7            metapod_1.16.0
[155] dqrng_0.4.1                survMisc_0.5.6
[157] R.utils_2.13.0             lazyeval_0.2.2
[159] shiny_1.11.1               htmltools_0.5.8.1
[161] KMsurv_0.1-6               ensembldb_2.32.0
[163] glue_1.8.0                 XVector_0.48.0
[165] RCurl_1.98-1.17            rprojroot_2.1.1
[167] coxme_2.2-22               jpeg_0.1-11
[169] gridExtra_2.3              AUCell_1.30.1
[171] boot_1.3-32                igraph_2.2.0
[173] R6_2.6.1                   gplots_3.2.0
[175] labeling_0.4.3             km.ci_0.5-6
[177] ggh4x_0.3.1                GenomicFeatures_1.60.0
[179] cluster_2.1.8.1            rngtools_1.5.2
[181] Rhdf5lib_1.30.0            nloptr_2.2.1
[183] DelayedArray_0.34.1        tidyselect_1.2.1
[185] vipor_0.4.7                ProtGenerics_1.40.0
[187] raster_3.6-32              ggforce_0.5.0
[189] car_3.1-3                  AnnotationDbi_1.70.0
[191] rsvd_1.0.5                 KernSmooth_2.23-26
[193] S7_0.2.0                   data.table_1.17.8
[195] htmlwidgets_1.6.4          RColorBrewer_1.1-3
[197] rlang_1.1.6                spatstat.sparse_3.1-0
[199] lmerTest_3.1-3             ggnewscale_0.5.2
[201] beeswarm_0.4.0            </code></pre>
</div>
</div>
</section><section class="level2"><h3 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h3>
<p>The authors thank all their colleagues, particularly at The University of Sydney, Sydney Precision Data Science and Charles Perkins Centre for their support and intellectual engagement. Special thanks to Ellis Patrick, Shila Ghazanfar, Andy Tran, Helen, and Daniel for guiding and supporting the building of this workshop.</p>
</section></section><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config);
    }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sydney Data Science Centre.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
